{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3a06a937-a46b-4bfb-9020-1819a910f4fa",
   "metadata": {},
   "source": [
    "This analysis was performed succesfully in \"CentOS Linux release 7.5.1804 (Core)\". If you use different version of Linux or OS, you may need to adjust it accordingly and we are not responsible for any liability or error.\n",
    "\n",
    "Most of the figures were post-processed using Adobe Illustrator to allow for easier adjustment of font-size, color, lines, etc.\n",
    "\n",
    "Other information and issue reporting, please refer to https://github.com/muharif/PFProgression_Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78279eda-031d-4b8e-a4a2-30b90986c83a",
   "metadata": {},
   "source": [
    "## Init"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc98e20f-c81f-4b48-a0be-c5c588ad1560",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import pandas as pd\n",
    "import rpy2,os,re, shutil\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import copy\n",
    "sns.set(font=\"Arial\")\n",
    "sns.set_style(\"white\")\n",
    "import matplotlib\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "from sklearn.decomposition import PCA as sklearnPCA\n",
    "os.environ['KMP_DUPLICATE_LIB_OK']='True'\n",
    "import pickle\n",
    "from statannot import add_stat_annotation\n",
    "from matplotlib.gridspec import GridSpec\n",
    "\n",
    "\n",
    "#rpy2\n",
    "import rpy2.robjects as robjects\n",
    "import rpy2.robjects.numpy2ri\n",
    "rpy2.robjects.numpy2ri.activate()\n",
    "from rpy2.robjects import pandas2ri\n",
    "pandas2ri.activate()\n",
    "import warnings\n",
    "from rpy2.rinterface import RRuntimeWarning\n",
    "warnings.filterwarnings(\"ignore\", category=RRuntimeWarning)\n",
    "\n",
    "from scipy.stats import zscore, spearmanr, ttest_ind, pearsonr,hypergeom, f_oneway\n",
    "from scipy.interpolate import interp1d\n",
    "from statsmodels.sandbox.stats.multicomp import multipletests\n",
    "import leidenalg\n",
    "import igraph as ig\n",
    "import gseapy as gp\n",
    "from scipy.spatial import distance\n",
    "from scipy.cluster.hierarchy import dendrogram, linkage, cut_tree\n",
    "from sklearn.impute import KNNImputer\n",
    "from scipy.io import loadmat\n",
    "import itertools"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c1af434-f4d9-4cf3-8a54-f1b5edb94c0f",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa0fed84-6beb-4827-96ad-2068f7f272ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "class DESeq_Piano:\n",
    "    def __init__(self,respath='.'):\n",
    "        if respath=='.':\n",
    "            print('Result Path is Not Defined, any result will be written in this directory')\n",
    "            self.respath=respath\n",
    "        elif os.path.isdir(respath):\n",
    "            print('Result Path Exists, please be careful of overwriting')\n",
    "            self.respath=respath\n",
    "        else:\n",
    "            print('Result Path Does Not Exists, creating new directory')\n",
    "            os.mkdir(respath)\n",
    "            self.respath=respath\n",
    "        self.__Init()\n",
    "        \n",
    "    def __Init(self):\n",
    "        robjects.r('''\n",
    "            library('DESeq2')\n",
    "            library('piano')\n",
    "            library('Biobase')\n",
    "            library('snow')\n",
    "            library('RColorBrewer')\n",
    "            library('gplots')\n",
    "            library('snowfall')\n",
    "        ''')\n",
    "        self.GSC_KEGG=None\n",
    "        self.GSC_TF=None\n",
    "        self.GSC_RM=None\n",
    "        self.GSC_GO=None\n",
    "        self.GSC_extraGSC=None\n",
    "        self.KEGG=None\n",
    "        self.TF=None\n",
    "        self.RM=None\n",
    "        self.GO=None\n",
    "        self.extraGSC=None\n",
    "        self.dds=None\n",
    "        self.deseq_res=None\n",
    "        self.conds=None\n",
    "        self.count=None\n",
    "        self.tpm=None\n",
    "\n",
    "    def __check_dir(self,extrapath):\n",
    "        if os.path.isdir('%s/%s/' % (self.respath,extrapath)):\n",
    "            print('Path %s/%s/ exist, writing in the folder' % (self.respath,extrapath))\n",
    "        else:\n",
    "            print('Path %s/%s/ does not exist, making the folder to write the results' % (self.respath,extrapath))\n",
    "            os.mkdir('%s/%s/' % (self.respath,extrapath))\n",
    "            \n",
    "    def DEseq(self,count_df=None,conds_series=None):\n",
    "        if count_df is not None:\n",
    "            self.count=count_df\n",
    "        if conds_series is not None:\n",
    "            self.conds=conds_series\n",
    "        self.gene_names=count_df.index\n",
    "        robjects.globalenv['conds']=np.array(self.conds)\n",
    "        robjects.globalenv['subject']=np.array(self.conds.index)\n",
    "        robjects.globalenv['count']=self.count.astype(int).values\n",
    "        robjects.globalenv['gene_names']=np.array(self.gene_names)\n",
    "        robjects.r('''\n",
    "            conds=as.factor(conds)\n",
    "            coldata <- data.frame(row.names=subject,conds)\n",
    "            dds <- DESeqDataSetFromMatrix(countData=as.matrix(count),colData=coldata,design=~conds)\n",
    "            dds <- DESeq(dds)\n",
    "            print(resultsNames(dds))\n",
    "            normalized_counts = counts(dds,normalized=TRUE)\n",
    "            rownames(normalized_counts) = gene_names\n",
    "            write.table(normalized_counts, file=\"../Data/normalized_counts.txt\", sep=\"\\t\", quote=F, col.names=NA)\n",
    "        ''')\n",
    "        self.dds=robjects.globalenv['dds']\n",
    "    \n",
    "    def DEseq_Compare(self, cond1, cond2,save=True):\n",
    "        if self.dds is None:\n",
    "            raise ValueError('No DESeq data found, run the DEseq function')\n",
    "        else:\n",
    "            print('Comparing %s and %s' % (cond1,cond2))\n",
    "            robjects.globalenv['dds']=self.dds\n",
    "            robjects.globalenv['cond1']=cond1\n",
    "            robjects.globalenv['cond2']=cond2\n",
    "            robjects.globalenv['gene_names']=np.array(self.gene_names)\n",
    "            robjects.r('''\n",
    "                res=results(dds,contrast=c('conds',cond1,cond2))\n",
    "                res=data.frame(res)\n",
    "                res$rownames=gene_names\n",
    "            ''')\n",
    "            res = pd.DataFrame(robjects.globalenv['res']).set_index('rownames')\n",
    "            self.deseq_df=res\n",
    "            if save:\n",
    "                self.__check_dir('deseq')\n",
    "                res.to_csv('%s/deseq/deseq_%s_%s.txt' % (self.respath,cond1,cond2),sep='\\t')\n",
    "        \n",
    "    def loadGSC(self,KEGG='',TF='',RM='',GO='',extraGSC=''):\n",
    "        if KEGG != '':\n",
    "            name='KEGG'\n",
    "            self.__check_dir(name)\n",
    "            self.GSC_KEGG=self.__loadGSC_execute(KEGG)\n",
    "        if GO != '':\n",
    "            name='GO'\n",
    "            self.__check_dir(name)\n",
    "            self.GSC_GO=self.__loadGSC_execute(GO)\n",
    "    \n",
    "    def __loadGSC_execute(self,GSC):\n",
    "        robjects.globalenv['GSC']=GSC \n",
    "        robjects.r('''\n",
    "            y=loadGSC(GSC)\n",
    "        ''')\n",
    "        return robjects.globalenv['y']\n",
    "    \n",
    "    def __PIANO_execute(self,cond1,cond2,GSCtype='',save=True):\n",
    "        robjects.globalenv['deseq_file']=pandas2ri.py2rpy_pandasdataframe(self.deseq_df)\n",
    "        if GSCtype == 'KEGG':\n",
    "            robjects.globalenv['y']=self.GSC_KEGG\n",
    "            deseq_df=self.deseq_df\n",
    "            deseq_df.index=[i.upper() for i in deseq_df.index]\n",
    "            robjects.globalenv['deseq_file']=pandas2ri.py2rpy_pandasdataframe(deseq_df)\n",
    "        elif GSCtype == 'GO':\n",
    "            robjects.globalenv['y']=self.GSC_GO\n",
    "            deseq_df=self.deseq_df\n",
    "            deseq_df.index=[i.upper() for i in deseq_df.index]\n",
    "            robjects.globalenv['deseq_file']=pandas2ri.py2rpy_pandasdataframe(deseq_df)\n",
    "        robjects.r('''\n",
    "            DESeq_file=deseq_file\n",
    "            DESeq_file=DESeq_file[ ,c('log2FoldChange','pvalue')]\n",
    "            logFC=as.matrix(DESeq_file[,1])\n",
    "            pval=as.matrix(DESeq_file[,2])\n",
    "            rownames(logFC)=(rownames(DESeq_file))\n",
    "            rownames(pval)=(rownames(DESeq_file))\n",
    "            logFC[is.na(logFC)] <- 0\n",
    "            pval[is.na(pval)] <- 1\n",
    "            gsaRes <- runGSA(pval,logFC,gsc=y, geneSetStat=\"reporter\", signifMethod=\"nullDist\", nPerm=1000,ncpus=8,gsSizeLim = c(5, Inf))\n",
    "            res_piano=GSAsummaryTable(gsaRes)\n",
    "            res_piano$rownames=rownames(res_piano)\n",
    "        ''')\n",
    "        res=pd.DataFrame(robjects.globalenv['res_piano']).set_index('rownames').set_index('Name').iloc[0:,0:]#.groupby('Name').sum()\n",
    "        if save:\n",
    "            res.to_csv('%s/%s/piano_%s_%s.txt' % (self.respath,GSCtype,cond1,cond2),sep='\\t')\n",
    "        return res\n",
    "\n",
    "    def PIANO(self,cond1,cond2,save=True):\n",
    "        if self.deseq_df is None:\n",
    "            raise ValueError('No DEseq comparison found, run DEseq_Compare first or assign deseq result dataframe to deseq_res')\n",
    "        else:\n",
    "            if self.GSC_KEGG is not None:\n",
    "                name='KEGG'\n",
    "                self.KEGG = self.__PIANO_execute(cond1,cond2,GSCtype=name,save=save)\n",
    "            if self.GSC_GO is not None:\n",
    "                name='GO'\n",
    "                self.GO = self.__PIANO_execute(cond1,cond2,GSCtype=name,save=save)         \n",
    "                    \n",
    "    def __piano2xlsx(self,name,part='ALL'):\n",
    "        if part != 'ALL':\n",
    "            files=[i for i in sorted(os.listdir('%s/%s' % (self.respath,name))) if part in i]\n",
    "        else:\n",
    "            files=sorted(os.listdir('%s/%s' % (self.respath,name)))\n",
    "        writer = pd.ExcelWriter('%s/%s_%s.xlsx' % (self.respath,name,part), engine='xlsxwriter')\n",
    "        for i in files:\n",
    "            if i.startswith('.'):\n",
    "                continue\n",
    "            deseq1=pd.read_csv('%s/%s/%s' % (self.respath,name,i),index_col='Name',sep='\\t')[['Genes (tot)','Stat (dist.dir.up)','p adj (dist.dir.up)']]\n",
    "            deseq1.columns=['# of Genes', 'Stats','P-Adj']\n",
    "            deseq1=deseq1[deseq1['P-Adj']<0.05]\n",
    "            deseq1['Direction']='UP'\n",
    "            deseq2=pd.read_csv('%s/%s/%s' % (self.respath,name,i),index_col='Name',sep='\\t')[['Genes (tot)','Stat (dist.dir.up)','p adj (dist.dir.dn)']]\n",
    "            deseq2.columns=['# of Genes', 'Stats','P-Adj']\n",
    "            deseq2=deseq2[deseq2['P-Adj']<0.05]\n",
    "            deseq2['Direction']='DOWN'\n",
    "            deseq1=pd.concat([deseq1,deseq2])\n",
    "            if len(i.replace('piano_','').replace('.txt',''))>31:\n",
    "                print('sheet name too big, trimming to 30 character')\n",
    "                n_temp=i.replace('piano_','').replace('.txt','')[0:31]\n",
    "            else:\n",
    "                n_temp=i.replace('piano_','').replace('.txt','')\n",
    "            deseq1.sort_values('P-Adj').to_excel(writer, sheet_name=n_temp)\n",
    "        writer.save()\n",
    "        \n",
    "    def summarizePiano(self,names=['KEGG','GO','TF','RM','extraGSC'],part='ALL'):\n",
    "        for name in names:\n",
    "            if os.path.isdir(self.respath+name):\n",
    "                self.__piano2xlsx(name,part=part)\n",
    "            else:\n",
    "                continue\n",
    "    \n",
    "    def summarizeDEseq(self,deseq_result='deseq',part='ALL'):\n",
    "        if os.path.isdir(self.respath+deseq_result):\n",
    "            if part != 'ALL':\n",
    "                files=[i for i in sorted(os.listdir(self.respath+'/deseq/')) if part in i]\n",
    "            else:\n",
    "                files=sorted(os.listdir(self.respath+'/deseq/'))\n",
    "            writer = pd.ExcelWriter(self.respath+'/DifferentialExpression_%s.xlsx' % part, engine='xlsxwriter')\n",
    "            fin=0\n",
    "            for i in files:\n",
    "                if i.startswith('.'):\n",
    "                    continue\n",
    "                deseq=pd.read_csv(self.respath+'/deseq/%s' % i,index_col=0,sep='\\t')\n",
    "                deseq['abs']=deseq['log2FoldChange'].abs()\n",
    "                deseq=deseq.sort_values('abs',ascending=False)[['log2FoldChange','pvalue','padj']]\n",
    "\n",
    "                deseq['Direction']=['UP' if i > 0 else 'DOWN' for i in deseq['log2FoldChange']]\n",
    "                deseq.columns=['L2FC','P-VALUE','P-ADJ','DIRECTION']\n",
    "                deseq=deseq.sort_values('P-VALUE')\n",
    "                if len(i.replace('deseq_','').replace('.txt',''))>31:\n",
    "                    print('sheet name too big, trimming to 30 character')\n",
    "                    n_temp=i.replace('deseq_','').replace('.txt','')[0:31]\n",
    "                else:\n",
    "                    n_temp=i.replace('deseq_','').replace('.txt','')\n",
    "                deseq.to_excel(writer, sheet_name=n_temp)\n",
    "            writer.save()\n",
    "        else:\n",
    "            raise ValueError('No DESEq result found')\n",
    "\n",
    "    def save_object(self,filename='DEseq_PIANO.pkl'):\n",
    "        with open('%s/%s' % (self.respath,filename), 'wb') as file:\n",
    "            pickle.dump(self, file)\n",
    "    \n",
    "    def load_object(self,filename='DEseq_PIANO.pkl'):\n",
    "        with open(filename, 'rb') as file:\n",
    "            res=pickle.load(file)\n",
    "        return res\n",
    "\n",
    "def filter_tpm(tpm,conds,thr=1):\n",
    "    #making sure that in at least 1 condition we have TPM > threshold (default 1)\n",
    "    return tpm[(pd.concat([tpm.T,conds],1).groupby(conds.name).mean().T>thr).sum(1)>0]\n",
    "\n",
    "def calc(tpm, clin = None,pval_thr=0.05,pos_only=False):\n",
    "    print('Calculating Correlation..')\n",
    "    if (clin) == None:\n",
    "        tpm = tpm\n",
    "    else:   \n",
    "        tpm = pd.concat([tpm.T,clin], 1).T\n",
    "        temp=spearmanr(tpm.T)\n",
    "        corr=pd.DataFrame(temp[0],columns=list(tpm.index),index=list(tpm.index))\n",
    "        pval=pd.DataFrame(temp[1],columns=list(tpm.index),index=list(tpm.index))\n",
    "        print('Filtering the matrix Correlation..')\n",
    "        corr=corr.where(np.triu(np.ones(corr.shape)).astype(np.bool))\n",
    "        pval=pval.where(np.triu(np.ones(pval.shape)).astype(np.bool))\n",
    "        print('Making long table of Correlation..')\n",
    "        corr2=corr.unstack().reset_index(name='weight')\n",
    "        pval2=pval.unstack().reset_index(name='pval')\n",
    "        res=corr2.merge(pval2,on=['level_0','level_1'])\n",
    "        res=res[res['level_0'] != res['level_1']]\n",
    "        res=res.dropna()\n",
    "        print('Adjusting P-val')\n",
    "        res['padj']=multipletests(res['pval'],method='fdr_bh')[1]\n",
    "        res=res[res.padj < pval_thr].reset_index(drop=True)\n",
    "        if pos_only:\n",
    "            res=res[res['weight']>0]\n",
    "        res=res[['level_0','level_1','weight']]\n",
    "        print('Done!!')\n",
    "    return res\n",
    "\n",
    "\n",
    "class Network_Analysis:\n",
    "    def __init__(self,filename, network_type,respath,KEGG='',GO='',topx=0.05,cluster_size=30):\n",
    "        self.network_ori=pd.read_csv(filename,sep='\\t')\n",
    "        self.res_path=respath\n",
    "        if os.path.isdir(self.res_path):\n",
    "            pass\n",
    "        else:\n",
    "            os.mkdir(self.res_path)\n",
    "        self.cluster_size=cluster_size\n",
    "        self.topx=topx\n",
    "        if network_type == 'transcriptomics':\n",
    "            self.__net_analysis_transcriptomics()\n",
    "        else:\n",
    "            self.__net_analysis_otheromics()\n",
    "        \n",
    "    def __net_analysis_transcriptomics(self):\n",
    "        print('Loading The Network...')\n",
    "        temp=self.network_ori[self.network_ori['weight']>0]\n",
    "        temp=temp[temp['weight']>temp['weight'].quantile(1-self.topx)]\n",
    "        g= ig.Graph.TupleList(zip(temp['level_0'],temp['level_1']))\n",
    "        print('Cluster Analysis...')\n",
    "        optimiser = leidenalg.Optimiser()\n",
    "        clust_calc = leidenalg.ModularityVertexPartition(g)\n",
    "        diff =  optimiser.optimise_partition(clust_calc, n_iterations=-1)\n",
    "        clust=pd.Series(clust_calc.membership,index=g.vs['name'])\n",
    "        thr=self.cluster_size\n",
    "        temp_c=clust.value_counts()[clust.value_counts()>thr].index.tolist()\n",
    "        self.degree = pd.Series(g.degree(), index = g.vs['name'], name = 'degree')\n",
    "        self.clustering=clust[clust.isin(temp_c)]\n",
    "        self.modularity=clust_calc.modularity\n",
    "        self.network=g\n",
    "        temp=self.network_ori.merge(pd.DataFrame(self.clustering).reset_index(),how='left',left_on='level_0',right_on='index').merge(pd.DataFrame(self.clustering).reset_index(),how='left',left_on='level_1',right_on='index').dropna()\n",
    "        temp2=temp[['0_x','0_y']]\n",
    "        temp2['count']=1\n",
    "        temp2=temp2.groupby(['0_x','0_y']).count().reset_index().pivot_table(index='0_x',columns='0_y',values='count').fillna(0)\n",
    "        temp2=pd.DataFrame([(i,j,(temp2.loc[i,j]+temp2.loc[j,i]),(temp2.loc[i,j]+temp2.loc[j,i])/(self.clustering.value_counts()[i]+self.clustering.value_counts()[j])) for i in temp2.index for j in temp2.columns if (i < j)],columns=['Source','Target','Weight1','Weight2'])\n",
    "        temp2.astype(int).to_csv('%s/edge_pos_cluster.txt' % self.res_path,sep='\\t')\n",
    "        pd.DataFrame(self.clustering.value_counts(),columns=['Size']).to_csv('%s/node_pos_cluster.txt' % self.res_path,sep='\\t')\n",
    "        pd.concat([self.clustering.rename('cluster'),  self.degree.reindex(self.clustering.index)], 1).to_csv('%s/clust_pos_cluster.txt' % self.res_path,sep='\\t')\n",
    "\n",
    "        \n",
    "    def __net_analysis_otheromics(self):\n",
    "        print('Loading The Network...')\n",
    "        temp=self.network_ori\n",
    "        temp=temp[(temp['weight']>temp['weight'].quantile(1-self.topx)) | (temp['weight']<temp['weight'].quantile(self.topx))]\n",
    "        g= ig.Graph.TupleList(zip(temp['level_0'],temp['level_1'],temp['weight']),weights=True)\n",
    "        print('Cluster Analysis...')\n",
    "        G_pos = g.subgraph_edges(g.es.select(weight_gt = 0), delete_vertices=False)\n",
    "        G_neg = g.subgraph_edges(g.es.select(weight_lt = 0), delete_vertices=False)\n",
    "        G_neg.es['weight'] = [-w for w in G_neg.es['weight']]\n",
    "        part_pos = leidenalg.ModularityVertexPartition(G_pos, weights='weight')\n",
    "        part_neg = leidenalg.ModularityVertexPartition(G_neg, weights='weight');\n",
    "        optimiser = leidenalg.Optimiser()\n",
    "        diff = optimiser.optimise_partition_multiplex([part_pos, part_neg],layer_weights=[1,-1], n_iterations=-1)\n",
    "        clust = pd.DataFrame(pd.Series(part_pos.membership+part_neg.membership,index=G_pos.vs['name']+G_neg.vs['name'])).reset_index().drop_duplicates().set_index('index')[0]\n",
    "        thr=self.cluster_size\n",
    "        temp_c=clust.value_counts()[clust.value_counts()>thr].index.tolist()\n",
    "        self.degree_combi = pd.Series(g.degree(), index = g.vs['name'], name = 'degree')\n",
    "        self.clustering_combi=clust[clust.isin(temp_c)]\n",
    "        self.modularity_combi=diff\n",
    "        self.network_combi=g\n",
    "        \n",
    "        temp=self.network_ori.merge(pd.DataFrame(self.clustering_combi).reset_index(),how='left',left_on='level_0',right_on='index').merge(pd.DataFrame(self.clustering_combi).reset_index(),how='left',left_on='level_1',right_on='index').dropna()\n",
    "        temp2=temp[['0_x','0_y']]\n",
    "        temp2['count'] = 1\n",
    "        temp2=temp2.groupby(['0_x','0_y']).count().reset_index().pivot_table(index='0_x',columns='0_y',values='count').fillna(0)\n",
    "        temp2=pd.DataFrame([(i,j,(temp2.loc[i,j]+temp2.loc[j,i]),(temp2.loc[i,j]+temp2.loc[j,i])/(self.clustering_combi.value_counts()[i]+self.clustering_combi.value_counts()[j])) for i in temp2.index for j in temp2.columns if (i < j)],columns=['Source','Target','Weight1','Weight2'])\n",
    "        temp2.astype(int).to_csv('%s/edge_combi_cluster.txt' % self.res_path,sep='\\t')\n",
    "        pd.DataFrame(self.clustering_combi.rename('Size').value_counts(),columns=['Size']).to_csv('%s/node_combi_cluster.txt' % self.res_path,sep='\\t')\n",
    "        pd.concat([self.clustering_combi.rename('cluster'), self.degree_combi.reindex(self.clustering_combi.index)], 1).to_csv('%s/clust_combi_cluster.txt' % self.res_path,sep='\\t')\n",
    "\n",
    "    \n",
    "    def save_network(self):\n",
    "        pickle_out = open('%s/network_object.pkl' % self.res_path,\"wb\")\n",
    "        pickle.dump(self, pickle_out)\n",
    "        pickle_out.close()\n",
    "        \n",
    "def impute_knn(data):\n",
    "    imputer = KNNImputer(n_neighbors=2)\n",
    "    After_imputation = imputer.fit_transform(data)\n",
    "    return pd.DataFrame(After_imputation, index = data.index, columns = data.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a2c4c10-3534-40a8-89e3-c88db99e8709",
   "metadata": {},
   "source": [
    "## Data Loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ba67a73-b736-4b66-91b4-75dd468c861d",
   "metadata": {},
   "outputs": [],
   "source": [
    "supp1 = pd.ExcelFile('SupplementaryTables/Supplementary Table 1.xlsx') # Metadata, PFT, Ashcroft, WEight, Transcriptomics\n",
    "supp3 = pd.ExcelFile('SupplementaryTables/Supplementary Table 3.xlsx') # Metabolomics\n",
    "supp5 = pd.ExcelFile('SupplementaryTables/Supplementary Table 5.xlsx') # Human Transcriptomics\n",
    "supp6 = pd.ExcelFile('SupplementaryTables/Supplementary Table 6.xlsx') # iDREM\n",
    "supp7 = pd.ExcelFile('SupplementaryTables/Supplementary Table 7.xlsx') # CB1R"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a261db8c-2fc0-4389-8134-fe063092151f",
   "metadata": {},
   "source": [
    "## Figure 1B\n",
    "\n",
    "Created with GraphPad/Prism"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6e6c1bf-efcc-4a91-b7f4-43978f35b69d",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = supp1.parse('Metadata and PFTs', index_col = 0)\n",
    "metadata = temp.iloc[0:, 0:2]\n",
    "metadata['Conditions'] = [\"%s (%s)\" % (metadata['Treatment'][i], metadata['Day'][i]) if metadata['Treatment'][i] != 'Control' else 'Control' for i in metadata.index]\n",
    "weight = supp1.parse('BodyWeight', index_col = 0)\n",
    "weight = (weight.T/weight['Day-0']).T*100\n",
    "weight.columns = [int(i.split('-')[1]) for i in weight.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21363514-8c22-4596-bb80-dc1179df4487",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = weight.unstack().reset_index().dropna()\n",
    "plt.figure(figsize = (3, 3))\n",
    "g = sns.pointplot(data = temp, x = 'level_0', y = 0, aspect=5, color = 'grey', errorbar = 'se', errwidth= 1, capsize=.5,scale = 0.2)\n",
    "xticks = [i for i in weight.columns if i % 7 == 0]\n",
    "g.set_xticks(xticks)\n",
    "g.set_xticklabels(xticks, fontsize = 15)#, rotation = 30, ha = 'right')\n",
    "g.set_yticklabels(['%d' % j if int(j) == j else '%.2f' % j for j in g.get_yticks()], fontsize = 15)\n",
    "g.set_xlabel('Days\\nPost-Bleomycin', fontsize = 15, fontweight = 'bold')\n",
    "g.set_xlim(0, 29)\n",
    "g.set_ylabel('')\n",
    "plt.savefig('Figures/Figure1B.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "accd0216-8613-4ad5-92cd-0159466a5065",
   "metadata": {},
   "source": [
    "## Figure 1C-G"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14859446-d44f-491c-84d4-1f5d9b96a3d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = supp1.parse('Metadata and PFTs', index_col = 0)\n",
    "metadata = temp.iloc[0:, 0:2]\n",
    "metadata['Conditions'] = [\"%s (%s)\" % (metadata['Treatment'][i], metadata['Day'][i]) if metadata['Treatment'][i] != 'Control' else 'Control' for i in metadata.index]\n",
    "PFT = temp.iloc[0:, 2:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e85c80ee-4380-471e-ab40-722aa6b78e59",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(figsize = (10, 6),nrows = 2, ncols = 3, sharex=True)\n",
    "axs = sum([list(i) for i in axs], [])\n",
    "\n",
    "for num, pft in enumerate(PFT.columns):\n",
    "    temp = pd.concat([PFT[pft], metadata],1).reset_index()\n",
    "    order = ['Control','Bleomycin (7 Days)', 'Bleomycin (14 Days)','Bleomycin (21 Days)', 'Bleomycin (28 Days)']\n",
    "    box_pairs = [\n",
    "        ('Control', 'Bleomycin (7 Days)'),\n",
    "        ('Control', 'Bleomycin (14 Days)'),\n",
    "        ('Control', 'Bleomycin (21 Days)'),\n",
    "        ('Control', 'Bleomycin (28 Days)')\n",
    "                ]\n",
    "    \n",
    "    pval_sel = [\n",
    "        f_oneway(temp[temp['Conditions'] == 'Bleomycin (7 Days)'][pft], \n",
    "             temp[temp['Conditions'] == 'Control'][pft])[1],\n",
    "        f_oneway(temp[temp['Conditions'] == 'Bleomycin (14 Days)'][pft], \n",
    "             temp[temp['Conditions'] == 'Control'][pft])[1],\n",
    "        f_oneway(temp[temp['Conditions'] == 'Bleomycin (21 Days)'][pft], \n",
    "             temp[temp['Conditions'] == 'Control'][pft])[1],\n",
    "        f_oneway(temp[temp['Conditions'] == 'Bleomycin (28 Days)'][pft], \n",
    "             temp[temp['Conditions'] == 'Control'][pft])[1],\n",
    "    ]\n",
    "    \n",
    "    sns.boxplot(data = temp, x = 'Conditions', y = pft, order=order, ax = axs[num])\n",
    "    add_stat_annotation(data = temp, x = 'Conditions', y = pft, order=order, ax = axs[num],\n",
    "                        box_pairs=box_pairs,\n",
    "                        perform_stat_test=False, pvalues=pval_sel,\n",
    "                        test=None, text_format='star', loc='inside', verbose=0, fontsize = 15)\n",
    "    axs[num].set_xticklabels(axs[num].get_xticklabels(), fontsize = 15, fontweight = 'bold', rotation = 30, ha = 'right')\n",
    "    axs[num].set_yticklabels(['%d' % j if int(j) == j else '%.2f' % j for j in axs[num].get_yticks()], fontsize = 15)\n",
    "    axs[num].set_xlabel('')\n",
    "    axs[num].set_ylabel('')\n",
    "    axs[num].set_title('%s' % (pft.replace(' (PFT)', '')), fontsize = 15, fontweight = 'bold')\n",
    "    plt.subplots_adjust(wspace = 0.3)\n",
    "plt.savefig('Figures/Figure1C.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb0ea50b-de83-423c-9c96-b74cf2d05e39",
   "metadata": {},
   "source": [
    "## Figure 1H"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fa45bc6-e900-4116-b928-1440a61696fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "ashcroft = supp1.parse('Ashcroft', index_col = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2b4465b-07c5-44d6-9437-f1ed0dbbebad",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize = (3, 2.5))\n",
    "order = ['Control','Bleomycin (7 Days)', 'Bleomycin (14 Days)','Bleomycin (21 Days)', 'Bleomycin (28 Days)']\n",
    "box_pairs = [\n",
    "    ('Control', 'Bleomycin (7 Days)'),\n",
    "    ('Control', 'Bleomycin (14 Days)'),\n",
    "    ('Control', 'Bleomycin (21 Days)'),\n",
    "    ('Control', 'Bleomycin (28 Days)')\n",
    "            ]\n",
    "\n",
    "pval_sel = [\n",
    "    f_oneway(ashcroft[ashcroft['Group'] == 'Bleomycin (7 Days)']['Score'], \n",
    "         ashcroft[ashcroft['Group'] == 'Control']['Score'])[1],\n",
    "    f_oneway(ashcroft[ashcroft['Group'] == 'Bleomycin (14 Days)']['Score'], \n",
    "         ashcroft[ashcroft['Group'] == 'Control']['Score'])[1],\n",
    "    f_oneway(ashcroft[ashcroft['Group'] == 'Bleomycin (21 Days)']['Score'], \n",
    "         ashcroft[ashcroft['Group'] == 'Control']['Score'])[1],\n",
    "    f_oneway(ashcroft[ashcroft['Group'] == 'Bleomycin (28 Days)']['Score'], \n",
    "         ashcroft[ashcroft['Group'] == 'Control']['Score'])[1],\n",
    "]\n",
    "\n",
    "sns.boxplot(data = ashcroft, x = 'Group', y = 'Score', order=order, ax = ax)\n",
    "# sns.stripplot(data = ashcroft, x = 'Group', y = 'Score', order=order, marker = 's', color = 'black', alpha = 0.25, ax = ax)\n",
    "add_stat_annotation(data = ashcroft, x = 'Group', y = 'Score', order=order, ax = ax,\n",
    "                    box_pairs=box_pairs,\n",
    "                    perform_stat_test=False, pvalues=pval_sel,\n",
    "                    test=None, text_format='star', loc='inside', verbose=0, fontsize = 15)\n",
    "ax.set_xticklabels(ax.get_xticklabels(), fontsize = 15, fontweight = 'bold', rotation = 30, ha = 'right')\n",
    "ax.set_yticklabels(['%d' % j if int(j) == j else '%.2f' % j for j in ax.get_yticks()], fontsize = 15)\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('')\n",
    "ax.set_ylim([0, 8])\n",
    "ax.set_title('Ashcroft Score', fontsize = 15, fontweight = 'bold')\n",
    "plt.subplots_adjust(wspace = 0.3)\n",
    "plt.savefig('Figures/Figure1D.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c76e8221-e34e-446f-bae8-b2a24a2cbf70",
   "metadata": {},
   "source": [
    "## Supplementary Figure 1set_ylim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c282e97d-7936-4d42-abf7-1ebd554f48e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = supp1.parse('Metadata and PFTs', index_col = 0)\n",
    "metadata = temp.iloc[0:, 0:2]\n",
    "metadata['Conditions'] = [\"%s (%s)\" % (metadata['Treatment'][i], metadata['Day'][i]) if metadata['Treatment'][i] != 'Control' else 'Control' for i in metadata.index]\n",
    "qpcr = supp1.parse('qPCR_FibrosisMarkers', index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "235ecd2d-0f8a-4cc1-9c56-4cc8aedccbf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(figsize = (10, 12),nrows = 2, ncols = 2, sharex=True)\n",
    "axs = axs.flatten()\n",
    "for ax,gene in list(zip(axs, qpcr.columns)):\n",
    "    fc = pd.concat([qpcr[gene], metadata],1).dropna()\n",
    "    order = ['Control','Bleomycin (7 Days)', 'Bleomycin (14 Days)','Bleomycin (21 Days)', 'Bleomycin (28 Days)']\n",
    "    box_pairs = [\n",
    "        ('Control', 'Bleomycin (7 Days)'),\n",
    "        ('Control', 'Bleomycin (14 Days)'),\n",
    "        ('Control', 'Bleomycin (21 Days)'),\n",
    "        ('Control', 'Bleomycin (28 Days)')\n",
    "                ]\n",
    "    \n",
    "    pval_sel = [\n",
    "        f_oneway(fc[fc['Conditions'] == 'Bleomycin (7 Days)'][gene], \n",
    "             fc[fc['Conditions'] == 'Control'][gene])[1],\n",
    "        f_oneway(fc[fc['Conditions'] == 'Bleomycin (14 Days)'][gene], \n",
    "             fc[fc['Conditions'] == 'Control'][gene])[1],\n",
    "        f_oneway(fc[fc['Conditions'] == 'Bleomycin (21 Days)'][gene], \n",
    "             fc[fc['Conditions'] == 'Control'][gene])[1],\n",
    "        f_oneway(fc[fc['Conditions'] == 'Bleomycin (28 Days)'][gene], \n",
    "             fc[fc['Conditions'] == 'Control'][gene])[1],\n",
    "    ]\n",
    "    \n",
    "    sns.barplot(data = fc, x = 'Conditions', y = gene, order=order, capsize=.2, ax = ax)\n",
    "    sns.stripplot(data = fc, x = 'Conditions', y = gene, order=order, marker = 's', color = 'black', alpha = 0.75, ax = ax)\n",
    "    add_stat_annotation(data = fc, x = 'Conditions', y = gene, order=order, ax = ax,\n",
    "                        box_pairs=box_pairs,\n",
    "                        perform_stat_test=False, pvalues=pval_sel,\n",
    "                        test=None, text_format='star', loc='inside', verbose=0, fontsize = 15)\n",
    "    ax.set_title('%s' % gene, fontsize = 15, fontweight = 'bold')\n",
    "    ax.set_xticklabels(ax.get_xticklabels(), fontsize = 15, fontweight = 'bold', rotation = 30, ha = 'right')\n",
    "    ax.set_yticklabels(['%d' % j if int(j) == j else '%.1f' % j for j in ax.get_yticks()], fontsize = 15)\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_ylabel('')\n",
    "    plt.tight_layout()\n",
    "plt.savefig('Figures/Supp1.pdf', transparent = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8dcdad86-b338-4933-8354-edf249d333bd",
   "metadata": {},
   "source": [
    "## Figure 2A\n",
    "\n",
    "Figure 2A was rotated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe4cf333-da48-460f-8b81-b5c908b8d82a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "temp = supp1.parse('Metadata and PFTs', index_col = 0)\n",
    "metadata = temp.iloc[0:, 0:2]\n",
    "metadata['Conditions'] = [\"%s (%s)\" % (metadata['Treatment'][i], metadata['Day'][i]) if metadata['Treatment'][i] != 'Control' else 'Control' for i in metadata.index]\n",
    "tpm = supp1.parse('TPM', index_col = 0)[metadata.index]\n",
    "sel = filter_tpm(tpm, metadata['Conditions'], thr = 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c55e44d0-c0de-4b9f-a00d-725fb64e1e94",
   "metadata": {},
   "outputs": [],
   "source": [
    "var = 'Conditions'\n",
    "sklearn_pca = sklearnPCA(n_components=2)\n",
    "Y_sklearn = sklearn_pca.fit_transform(np.log1p(sel).fillna(0).T)\n",
    "plt.figure(figsize=(6,5))\n",
    "coordinate = pd.DataFrame(Y_sklearn, columns = ['x', 'y'], index = sel.columns) * -1 ## -1 is to rotate the PCA, so Control will be on the left, for aesthetic purpose\n",
    "coordinate = pd.concat([coordinate, metadata],1)\n",
    "ax = sns.scatterplot(data = coordinate, x = 'x', y = 'y', s = 300, hue = var, palette = 'colorblind')\n",
    "ax.set_xlabel('PC1 (%.2f%%)' % (sklearn_pca.explained_variance_ratio_[0]*100), fontsize = 20, fontweight = 'bold')\n",
    "ax.set_ylabel('PC2 (%.2f%%)' % (sklearn_pca.explained_variance_ratio_[1]*100), fontsize = 20, fontweight = 'bold')\n",
    "ax.set_xticklabels([])\n",
    "ax.set_yticklabels([])\n",
    "ax.get_legend().set_title('Groups')\n",
    "plt.setp(ax.get_legend().get_title(), fontsize='15', fontweight = 'bold')\n",
    "plt.setp(ax.get_legend().get_texts(), fontsize='15')\n",
    "plt.tight_layout()\n",
    "plt.legend(bbox_to_anchor=(1.05, 1.0), loc='upper left')\n",
    "# plt.savefig('Figures/Figure2A.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8dc7fc55-3cdc-49e6-b866-c4eaec1dc023",
   "metadata": {},
   "source": [
    "## Differential Expression Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c8dd180-bbf8-44f2-b646-bc96e3fcdb19",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = supp1.parse('Metadata and PFTs', index_col = 0)\n",
    "metadata = temp.iloc[0:, 0:2]\n",
    "metadata['Conditions'] = [\"%s (%s)\" % (metadata['Treatment'][i], metadata['Day'][i]) if metadata['Treatment'][i] != 'Control' else 'Control' for i in metadata.index]\n",
    "count = supp1.parse('Count', index_col = 0)[metadata.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f0a829f-bdcb-44de-96a6-4de490ef547d",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_path='StatisticalAnalysis/Transcriptomics_TimeSeries/'\n",
    "k=DESeq_Piano(res_path)\n",
    "k.DEseq(count,metadata['Conditions'].str.replace('\\(', '_').str.replace(')', '').str.replace(' ', ''))\n",
    "k.loadGSC(KEGG='lib/KEGG_2019_Mouse.gmt')\n",
    "\n",
    "cond2='Control'\n",
    "cond1='Bleomycin_7Days'\n",
    "k.DEseq_Compare(cond1,cond2,save=True)\n",
    "\n",
    "cond2='Control'\n",
    "cond1='Bleomycin_14Days'\n",
    "k.DEseq_Compare(cond1,cond2,save=True)\n",
    "k.PIANO(cond1,cond2,save=True)\n",
    "\n",
    "cond2='Control'\n",
    "cond1='Bleomycin_21Days'\n",
    "k.DEseq_Compare(cond1,cond2,save=True)\n",
    "\n",
    "cond2='Control'\n",
    "cond1='Bleomycin_28Days'\n",
    "k.DEseq_Compare(cond1,cond2,save=True)\n",
    "\n",
    "k.summarizeDEseq()\n",
    "k.summarizePiano()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15d03a6f-31d4-42d6-a747-ae4a8121a387",
   "metadata": {},
   "source": [
    "## Figure 2B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7458dcf4-352d-4b62-b740-a0d9db4f0507",
   "metadata": {},
   "outputs": [],
   "source": [
    "deseq = pd.ExcelFile('StatisticalAnalysis/Transcriptomics_TimeSeries/DifferentialExpression_ALL.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87216cfa-67e2-43e7-a818-21960fe8970c",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = pd.DataFrame()\n",
    "for i in deseq.sheet_names:\n",
    "    temp = deseq.parse(i, index_col = 0)\n",
    "    temp = temp[temp['P-ADJ'] < 0.05]['DIRECTION'].rename(i)\n",
    "    res = pd.concat([res, temp],1)\n",
    "    \n",
    "res = res[[\n",
    "    'Bleomycin_7Days_Control',\n",
    "    'Bleomycin_14Days_Control',\n",
    "    'Bleomycin_21Days_Control',\n",
    "    'Bleomycin_28Days_Control'\n",
    "]]\n",
    "\n",
    "res.shape\n",
    "temp = res.unstack().reset_index().dropna().groupby(['level_0',0]).count().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c94b09db-6536-4b5e-9180-a52c6c19d011",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize = (5,5))\n",
    "g = sns.barplot(data = temp, x = 'level_0', y = 'level_1', hue = 0, order = res.columns, hue_order = ['UP', 'DOWN'])\n",
    "g.set_yticklabels([int(i) for i in g.get_yticks()], fontsize = 15)\n",
    "g.set_xticklabels(g.get_xticklabels(),rotation=30, ha = 'right', fontsize = 15, fontweight = 'bold')\n",
    "g.set_xlabel('')\n",
    "g.set_ylabel('# of DEGs', fontsize= 20, fontweight = 'bold')\n",
    "g.get_legend().set_title('Direction')\n",
    "plt.setp(g.get_legend().get_title(), fontsize='15', fontweight = 'bold')\n",
    "plt.setp(g.get_legend().get_texts(), fontsize='15')\n",
    "plt.tight_layout()\n",
    "plt.savefig('Figures/Figure2B.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73e1f1b9-3400-4d2b-8b64-40608a40425b",
   "metadata": {},
   "source": [
    "## Figure 2C-E"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8a7c332-bcc5-4fe6-a3e9-f6edc3b614c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "deseq = pd.ExcelFile('StatisticalAnalysis/Transcriptomics_TimeSeries/DifferentialExpression_ALL.xlsx')\n",
    "\n",
    "res = pd.DataFrame()\n",
    "for i in deseq.sheet_names:\n",
    "    temp = deseq.parse(i, index_col = 0)\n",
    "    temp = temp[temp['P-ADJ'] < 0.05]['DIRECTION'].rename(i)\n",
    "    res = pd.concat([res, temp],1)\n",
    "    \n",
    "res = res[[\n",
    "    'Bleomycin_7Days_Control',\n",
    "    'Bleomycin_14Days_Control',\n",
    "    'Bleomycin_21Days_Control',\n",
    "    'Bleomycin_28Days_Control'\n",
    "]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dba6170f-d0fb-4a42-b1cb-d055133acd79",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = supp1.parse('Metadata and PFTs', index_col = 0)\n",
    "metadata = temp.iloc[0:, 0:2]\n",
    "metadata['Conditions'] = [\"%s (%s)\" % (metadata['Treatment'][i], metadata['Day'][i]) if metadata['Treatment'][i] != 'Control' else 'Control' for i in metadata.index]\n",
    "tpm = supp1.parse('TPM', index_col = 0)[metadata.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b39e542f-1954-4def-8fe4-883ddde5731b",
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_genes = res.dropna(thresh = 4)\n",
    "normalized_tpm = zscore((tpm.reindex(selected_genes.index)).T).T.rename(columns = metadata['Conditions'].to_dict())\n",
    "\n",
    "row_linkage = linkage(\n",
    "    distance.pdist(normalized_tpm.fillna(0)), method='average')\n",
    "clust = pd.Series([i[0] for i in cut_tree(row_linkage, n_clusters = 2)], index = normalized_tpm.index).sort_values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32be0f8b-a247-4f37-8a57-90e61da3a680",
   "metadata": {},
   "outputs": [],
   "source": [
    "vmin = normalized_tpm.min().min()\n",
    "vmax = normalized_tpm.max().max()\n",
    "order = ['Control', 'Bleomycin (7 Days)', 'Bleomycin (14 Days)', 'Bleomycin (21 Days)', 'Bleomycin (28 Days)']\n",
    "g = sns.clustermap(normalized_tpm.reindex(clust.index).fillna(0)[order], figsize = (5,8), yticklabels = 0, xticklabels = 1, center = 0, cmap = 'RdYlBu_r', col_cluster = False, row_cluster = False, row_linkage = row_linkage, vmin = vmin, vmax = vmax)\n",
    "g.ax_heatmap.vlines(metadata['Conditions'].value_counts()[order].cumsum(), ymin = g.ax_heatmap.get_ylim()[1], ymax = g.ax_heatmap.get_ylim()[0], color = 'k')\n",
    "g.ax_heatmap.hlines(clust.value_counts().sort_index(ascending = True).cumsum().tolist(), xmin = g.ax_heatmap.get_xlim()[1], xmax = g.ax_heatmap.get_xlim()[0], color = 'k')\n",
    "g.ax_heatmap.tick_params(right=False, bottom=False)\n",
    "xticklabels = metadata['Conditions'].value_counts().sort_index()[order]\n",
    "g.ax_heatmap.set_xticks(xticklabels.cumsum()-((xticklabels/2)))\n",
    "g.ax_heatmap.set_xticklabels(xticklabels.index.tolist(), fontsize = 15, fontweight = 'bold', rotation = 30, ha = 'right')\n",
    "g.ax_heatmap.set_title('DEG in ALL comparisons vs Control\\n(%d Genes)' % normalized_tpm.shape[0], fontsize = 20, fontweight = 'bold')\n",
    "g.savefig('Figures/Figure2C.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd1c8777-54f4-4d41-ad93-37ef41e99bdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(figsize = (3, 8), nrows = 2, ncols = 1)\n",
    "color_mapping = dict(zip(clust.value_counts().index,sns.color_palette(\"tab10\").as_hex()))\n",
    "\n",
    "for i, cl in enumerate(clust.value_counts().index):\n",
    "    days = {'A': 0, 'B': 7, 'C': 14, 'D': 21, 'E': 28}\n",
    "    x_interp = np.linspace(days['A'], days['E'], 100)\n",
    "    sel_col = ['Control', 'Bleomycin (7 Days)', 'Bleomycin (14 Days)','Bleomycin (21 Days)','Bleomycin (28 Days)']\n",
    "    mapping = dict(zip(sel_col, ['A', 'B', 'C', 'D', 'E']))\n",
    "    sel_data = normalized_tpm[sel_col].reindex(clust[clust == cl].index).rename(columns = mapping).T.reset_index().groupby('index').mean().T[['A', 'B', 'C', 'D', 'E']].rename(columns = days).dropna()\n",
    "    y_interp = interp1d(list(days.values()), list(sel_data.values), kind = 'quadratic')(x_interp)\n",
    "    data_interp = pd.DataFrame(y_interp, columns = x_interp).unstack().reset_index().groupby('level_0').mean().reset_index()\n",
    "    data_interp[0] = data_interp[0] - data_interp.loc[0,0]\n",
    "    temp_max = data_interp.copy()[0]\n",
    "    g = sns.lineplot(data = data_interp, x = 'level_0', y = 0, ax = axs[i], color = color_mapping[cl], linewidth=5, zorder = 15)\n",
    "    \n",
    "    if cl == clust.value_counts().index.tolist()[-1]:\n",
    "        g.set_xlabel('Days', fontweight = 'bold', fontsize = 20)\n",
    "        g.set_xticks(list(days.values()))\n",
    "        g.set_xticklabels(list(days.values()), fontweight = 'bold', fontsize = 15)\n",
    "    else:\n",
    "        g.set_xlabel('')\n",
    "        g.set_xticks([])\n",
    "    temp_ylim = tuple(g.get_ylim())\n",
    "    g.vlines([7, 14, 21], ymin = g.get_ylim()[0], ymax = g.get_ylim()[1], color = 'lightgray', zorder = 10)\n",
    "    g.set_ylim(temp_ylim)\n",
    "    g.set_yticks([])\n",
    "    g.set_ylabel('')\n",
    "    g.set_xlim((0,28))\n",
    "    \n",
    "plt.savefig('Figures/Figure2C_trend.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c11a06f-0212-491c-ae84-c956a6ffe85d",
   "metadata": {},
   "outputs": [],
   "source": [
    "KEGG = {i.split('\\t')[0]:[j for j in i.rstrip('\\n').split('\\t')[2:] if j != ''] for i in open('lib/KEGG_2019_Mouse.gmt', 'r').readlines()}\n",
    "exclude_KEGG = ['Hepatitis B', 'Human T-cell leukemia virus 1 infection', 'Regulation of actin cytoskeleton', 'Measles',\n",
    "           'MicroRNAs in cancer', 'Pathways in cancer', 'Influenza A', 'Oocyte meiosis', 'Progesterone-mediated oocyte maturation',\n",
    "           'Ribosome biogenesis in eukaryotes', 'Amoebiasis', 'Small cell lung cancer', 'Epstein-Barr virus infection', 'Human papillomavirus infection',\n",
    "           'Pertussis', 'Staphylococcus aureus infection', 'Proteoglycans in cancer', 'Leishmaniasis', 'Legionellosis', 'Viral carcinogenesis', \n",
    "           'Chagas disease (American trypanosomiasis)', 'Toxoplasmosis', 'Malaria', 'Metabolism of xenobiotics by cytochrome P450', 'Chronic myeloid leukemia',\n",
    "           'Prion diseases', 'Inflammatory bowel disease (IBD)', 'Kaposi sarcoma-associated herpesvirus infection', 'Parkinson disease', 'Hepatitis C', 'Systemic lupus erythematosus',\n",
    "           'Tuberculosis', 'Hepatitis B', 'Measles', 'Dilated cardiomyopathy (DCM)', 'Hematopoietic cell lineage', 'Adrenergic signaling in cardiomyocytes',\n",
    "           'Epstein-Barr virus infection', 'Drug metabolism', 'Salivary secretion', 'Thermogenesis', 'Glioma', 'Human cytomegalovirus infection', 'Melanoma', 'Chemical carcinogenesis',\n",
    "            'African trypanosomiasis', 'Collecting duct acid secretion', 'African trypanosomiasis', 'Vascular smooth muscle contraction', 'Fluid shear stress and atherosclerosis',\n",
    "            'Rheumatoid arthritis', 'Allograft rejection', 'Long term depression'\n",
    "          ]\n",
    "\n",
    "exclude_GO = ['',\n",
    "             \n",
    "             ]\n",
    "\n",
    "color_mapping = dict(zip(sorted(clust.unique()),sns.color_palette(\"tab10\").as_hex()))\n",
    "\n",
    "\n",
    "fig, axs = plt.subplots(figsize = (3, 10),nrows = len(clust.unique().tolist()), ncols = 1, sharex=False\n",
    "                        , gridspec_kw={'height_ratios':[25, 6]})\n",
    "\n",
    "for j, sel_ax in enumerate(axs):\n",
    "    print()\n",
    "    gl = [ha.upper() for ha in clust[clust == j].index.tolist()]\n",
    "    enr = gp.enrichr(gene_list=gl,\n",
    "             gene_sets=KEGG,\n",
    "             background = tpm.shape[0],\n",
    "             outdir='test/enrichr_kegg2',\n",
    "             cutoff=1,\n",
    "             verbose=False)\n",
    "    temp = enr.res2d[~(\n",
    "        enr.res2d['Term'].isin(exclude_KEGG) |\n",
    "        (enr.res2d['Term'].str.contains('isease')) |\n",
    "         (enr.res2d['Term'].str.contains('arcinoma')) |\n",
    "          (enr.res2d['Term'].str.contains('ancer')) |\n",
    "        (enr.res2d['Term'].str.contains('thyroid')) |\n",
    "        (enr.res2d['Term'].str.contains('yndrome')) |\n",
    "        (enr.res2d['Term'].str.contains('strogen')) |\n",
    "        (enr.res2d['Term'].str.contains('elanoge')) |\n",
    "        (enr.res2d['Term'].str.contains('iabet')) |\n",
    "        (enr.res2d['Term'].str.contains('ysosome')) |\n",
    "        (enr.res2d['Term'].str.contains('nfection')) |\n",
    "        (enr.res2d['Term'].str.contains('ntestin')) |\n",
    "        (enr.res2d['Term'].str.contains('cardi')) |\n",
    "        (enr.res2d['Term'].str.contains('mmunodef'))\n",
    "    )].sort_values('Adjusted P-value', ascending = True)#.iloc[0:5]#.iloc[[2,3,4]]['Term'].tolist()\n",
    "    temp2 = temp[['Term','P-value','Adjusted P-value']]\n",
    "    temp2 = temp2[temp2['Adjusted P-value'] < 0.01]\n",
    "    print(temp2.shape)\n",
    "    temp2['Term'] = temp2['Term'].str.replace('endoplasmic reticulum', 'ER')\n",
    "    temp2['-Log10(P-Adj)'] = -np.log10(temp2['Adjusted P-value'])\n",
    "    sns.barplot(data = temp2, x = '-Log10(P-Adj)', y = 'Term', ax = sel_ax, color = color_mapping[j])\n",
    "    sel_ax.yaxis.tick_right()\n",
    "    sel_ax.set_yticklabels([i for i in temp2['Term']], fontsize = 15)\n",
    "#     sel_ax.set_xticks([])\n",
    "    sel_ax.set_ylabel('DEG-%d (%d Genes)' % (j, len(gl)), fontsize = 15, fontweight = 'bold')\n",
    "    sel_ax.tick_params(right = False)\n",
    "    if j == clust.max():\n",
    "        sns.despine(ax = sel_ax)\n",
    "        sel_ax.set_xlabel('-Log10(P-Adj)', fontsize = 15, fontweight = 'bold')\n",
    "    else:\n",
    "        sns.despine(ax = sel_ax, bottom = False)\n",
    "        sel_ax.set_xlabel('')\n",
    "plt.savefig('Figures/Figure2D-E.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29fd47df-b5f6-4cf7-bdb1-71e555996e5b",
   "metadata": {},
   "source": [
    "## Gene Co-Expression Network Generation and Analysis\n",
    "\n",
    "This analysis might take a long time, especially for the generation of the network. Pre-analyzed networks can be found under \"Networks/ReadyObjects\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "927ba30b-f0d8-485b-aabe-257eac10d86c",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = supp1.parse('Metadata and PFTs', index_col = 0)\n",
    "metadata = temp.iloc[0:, 0:2]\n",
    "metadata['Conditions'] = [\"%s (%s)\" % (metadata['Treatment'][i], metadata['Day'][i]) if metadata['Treatment'][i] != 'Control' else 'Control' for i in metadata.index]\n",
    "PFT = temp.iloc[0:, 2:]\n",
    "\n",
    "tpm = supp1.parse('TPM', index_col = 0)[metadata.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e108c20-6f35-461d-8af6-b2cf83c5e47d",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = pd.concat([tpm.T, metadata['Conditions']], 1).groupby('Conditions').mean()\n",
    "temp = temp.T[temp.std() > temp.std().quantile(.25)].index.tolist()\n",
    "tpm_filt = tpm.reindex(temp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "829a3ded-44fd-4724-a609-93bcd91c0133",
   "metadata": {},
   "outputs": [],
   "source": [
    "coexp = calc(tpm_filt, clin = PFT, pval_thr = 0.05)\n",
    "coexp.to_csv('Networks/GCN.txt', sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb8f2a37-c2f5-4394-bdf4-f4afa31a887a",
   "metadata": {},
   "outputs": [],
   "source": [
    "topx = 0.25\n",
    "k=Network_Analysis(filename='Networks/GCN.txt', network_type='transcriptomics',respath='Networks/GCN/',\n",
    "                   topx=topx, cluster_size = 30)\n",
    "k.save_network()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3871484-bf9e-45c2-86ae-6a22c77ef537",
   "metadata": {},
   "source": [
    "### Loading Pre-analyzed networks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88924e22-5f21-4012-97cb-a3a91433d5dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "k = pickle.load(open('Networks/ReadyObjects/GCN.pkl', 'rb'))\n",
    "\n",
    "# iGraph object is under k.network\n",
    "# Pandas object of the original network (not filtered based on the correlation threshold) is under k.network_ori\n",
    "# Subnetwork info iss under k.clustering\n",
    "# Degree centrality is under k.degree"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b3dab69-5938-47ea-9d3b-5ead7cb530a4",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Figure 3A\n",
    "\n",
    "Generated with cytoscape using the results of network analysis under \"Networks/GCN/\" or Supplementary Table 2. The size of the nodes was proportional to the average transitivity calculated below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96db5897-5f0e-4aef-9c0e-bd243921404c",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = []\n",
    "for cl in sorted(k.clustering.unique()):\n",
    "    k_temp = k.network.subgraph(k.clustering[k.clustering == cl].index)\n",
    "    print(cl)\n",
    "    print(k_temp.transitivity_avglocal_undirected())\n",
    "    x.append([cl, k_temp.transitivity_avglocal_undirected(), cl])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8776afc4-e44e-4a6e-baec-9a7c26075116",
   "metadata": {},
   "source": [
    "## Figure 3B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a097252-ae4f-4226-8618-05bdc1ca0e97",
   "metadata": {},
   "outputs": [],
   "source": [
    "PFT_Degree = k.degree[k.degree.index.str.contains('PFT')].sort_values(ascending = False).reset_index()\n",
    "\n",
    "sns.barplot(data = PFT_Degree, x = 'degree', y = 'index')\n",
    "plt.savefig('Figures/Figure3B.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe0833ae-9e2b-4fd0-a214-52d3e48f5469",
   "metadata": {},
   "source": [
    "## Figure 3C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a119fdd6-5faa-4fe6-b853-9869728c9e1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "k = pickle.load(open('Networks/ReadyObjects/GCN.pkl', 'rb'))\n",
    "\n",
    "clusters = pd.concat([k.clustering.rename('cluster'), k.degree],1)\n",
    "clusters = clusters[clusters['cluster'].notna()]\n",
    "clusters['norm_deg'] = ((clusters['degree'] - clusters['degree'].min())/(clusters['degree'].max() - clusters['degree'].min()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9f943d9-0ff9-487f-bf3f-ed9b7f11b0c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = supp1.parse('Metadata and PFTs', index_col = 0)\n",
    "metadata = temp.iloc[0:, 0:2]\n",
    "metadata['Conditions'] = [\"%s (%s)\" % (metadata['Treatment'][i], metadata['Day'][i]) if metadata['Treatment'][i] != 'Control' else 'Control' for i in metadata.index]\n",
    "PFT = temp.iloc[0:, 2:]\n",
    "\n",
    "tpm = supp1.parse('TPM', index_col = 0)[metadata.index]\n",
    "\n",
    "norm_tpm = zscore(tpm.T).T.rename(columns = metadata['Conditions'].to_dict())\n",
    "norm_pft = zscore(PFT).T.rename(columns = metadata['Conditions'].to_dict())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "386a6883-11ac-4bbf-a246-4e859d88be81",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = []\n",
    "for cl in sorted(clusters['cluster'].unique()):\n",
    "    sel_data = norm_tpm.reindex(clusters[clusters['cluster'] == cl].index).rename(columns = metadata['Conditions'].to_dict())#.T#.reset_index().groupby('index').mean().T.rename(columns = days)\n",
    "    row_linkage = linkage(distance.pdist(sel_data.fillna(0)), method='average')\n",
    "    dn = dendrogram(row_linkage, no_plot = True)['leaves']\n",
    "    temp = sel_data.iloc[dn].index.tolist()\n",
    "    x.append(temp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "387dcfa4-d916-4bf8-990a-488f128577ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "KEGG = {i.split('\\t')[0]:[j for j in i.rstrip('\\n').split('\\t')[2:] if j != ''] for i in open('lib/KEGG_2019_Mouse.gmt', 'r').readlines()}\n",
    "exclude_KEGG = ['Hepatitis B', 'Human T-cell leukemia virus 1 infection', 'Regulation of actin cytoskeleton', 'Measles',\n",
    "           'MicroRNAs in cancer', 'Pathways in cancer', 'Influenza A', 'Oocyte meiosis', 'Progesterone-mediated oocyte maturation',\n",
    "           'Ribosome biogenesis in eukaryotes', 'Amoebiasis', 'Small cell lung cancer', 'Epstein-Barr virus infection', 'Human papillomavirus infection',\n",
    "           'Pertussis', 'Staphylococcus aureus infection', 'Proteoglycans in cancer', 'Leishmaniasis', 'Legionellosis', 'Viral carcinogenesis', \n",
    "           'Chagas disease (American trypanosomiasis)', 'Toxoplasmosis', 'Malaria', 'Metabolism of xenobiotics by cytochrome P450', 'Chronic myeloid leukemia',\n",
    "           'Prion diseases', 'Inflammatory bowel disease (IBD)', 'Kaposi sarcoma-associated herpesvirus infection', 'Parkinson disease', 'Hepatitis C', 'Systemic lupus erythematosus',\n",
    "           'Tuberculosis', 'Hepatitis B', 'Measles', 'Dilated cardiomyopathy (DCM)', 'Hematopoietic cell lineage', 'Adrenergic signaling in cardiomyocytes',\n",
    "           'Epstein-Barr virus infection', 'Drug metabolism', 'Salivary secretion', 'Thermogenesis', 'Glioma', 'Human cytomegalovirus infection', 'Melanoma', 'Chemical carcinogenesis',\n",
    "            'African trypanosomiasis', 'Collecting duct acid secretion', 'African trypanosomiasis', 'Vascular smooth muscle contraction', 'Fluid shear stress and atherosclerosis',\n",
    "            'Rheumatoid arthritis', 'Allograft rejection', 'Long term depression'\n",
    "          ]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db1df30f-9ecf-4ba5-8861-e9d658dcef7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata['Conditions']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ec55fd1-2057-4046-ad83-1d07c55ce3b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "days = {'Control': 0, 'Bleomycin (7 Days)': 7, 'Bleomycin (14 Days)': 14, 'Bleomycin (21 Days)': 21, 'Bleomycin (28 Days)': 28}\n",
    "x_interp = np.linspace(days['Control'], days['Bleomycin (28 Days)'], 100)\n",
    "color_mapping = dict(zip(sorted(clusters['cluster'].unique()),sns.color_palette(\"tab10\").as_hex()[::-1]))\n",
    "\n",
    "cluster_color = pd.Series(clusters['cluster']).map(color_mapping)\n",
    "sel_data = norm_tpm.reindex(sum(x,[]))\n",
    "cm = sns.clustermap(sel_data.fillna(0), figsize=(12.63, 10),  yticklabels = 0, xticklabels = 0, center = 0, cmap = 'RdYlBu_r', row_colors = cluster_color, row_cluster = False, col_cluster = False)\n",
    "cm.cax.set_visible(False)\n",
    "cm.gs.update(left=0.05, right=0.25)\n",
    "cm.ax_heatmap.vlines(metadata.groupby('Conditions').count()['Treatment'].cumsum().tolist(), ymin = 0, ymax = sel_data.shape[0], color = 'gray')\n",
    "cm.ax_heatmap.hlines(np.cumsum([len(i) for i in x]), xmin = 0, xmax = sel_data.shape[1], color = 'black')\n",
    "cm.ax_heatmap.set_xlabel('Days', fontweight = 'bold', fontsize = 20)\n",
    "cm.ax_heatmap.set_ylabel('', fontweight = 'bold', fontsize = 20)\n",
    "cm.ax_heatmap.set_xticks((metadata.groupby('Conditions').count().cumsum() - metadata.groupby('Conditions').count()/2)['Treatment'].tolist())\n",
    "cm.ax_heatmap.set_xticklabels([0, 7, 14, 21, 28], fontweight = 'bold', fontsize = 15)\n",
    "\n",
    "gs = GridSpec(int(clusters['cluster'].max()+2),5, top = .95, left = 0.26, hspace = 0.2, \n",
    "              wspace = 0.21,width_ratios=[.5,0.1, .5,  1, 0.25])\n",
    "\n",
    "for cl in sorted(clusters['cluster'].unique().astype(int)):\n",
    "    sel_ax = cm.fig.add_subplot(gs[cl+1, 0])\n",
    "    sel_data = norm_tpm.reindex(\n",
    "        clusters[clusters['cluster'] == cl].index\n",
    "    ).rename(columns = metadata['Conditions'].to_dict()).T.reset_index().groupby('index').mean().T.rename(columns = days).dropna()[days.values()]\n",
    "    y_interp = interp1d(list(days.values()), list(sel_data.values), kind = 'quadratic')(x_interp)\n",
    "    data_interp = pd.DataFrame(y_interp, columns = x_interp).unstack().reset_index().groupby('level_0').mean().reset_index()\n",
    "    temp_max = data_interp.copy()[0]\n",
    "    g = sns.lineplot(data = data_interp, x = 'level_0', y = 0, ax = sel_ax, color = color_mapping[cl])\n",
    "    sel_data = norm_pft.reindex(\n",
    "        clusters[clusters['cluster'] == cl].index\n",
    "    ).rename(columns = metadata['Conditions'].to_dict()).T.reset_index().groupby('index').mean().T.rename(columns = days).dropna()[days.values()]\n",
    "    \n",
    "    if sel_data.shape[0] > 0:\n",
    "        y_interp = interp1d(list(days.values()), list(sel_data.values), kind = 'quadratic')(x_interp)\n",
    "        temp_max = pd.concat([temp_max, data_interp.copy()[0]])\n",
    "        data_interp = pd.DataFrame(y_interp, columns = x_interp).unstack().reset_index().groupby('level_0').mean().reset_index()\n",
    "        g = sns.lineplot(data = data_interp, x = 'level_0', y = 0, ax = sel_ax, color = color_mapping[cl], linestyle='--')\n",
    "    if cl == clusters['cluster'].max():\n",
    "        g.set_xlabel('Days', fontweight = 'bold', fontsize = 20)\n",
    "        g.set_xticks(list(days.values()))\n",
    "        g.set_xticklabels(list(days.values()), fontweight = 'bold', fontsize = 15)\n",
    "    else:\n",
    "        g.set_xlabel('')\n",
    "        g.set_xticks([])\n",
    "    temp_ylim = tuple(g.get_ylim())\n",
    "    g.vlines([7, 14, 21], ymin = g.get_ylim()[0], ymax = g.get_ylim()[1], color = 'lightgray')\n",
    "    g.set_ylim(temp_ylim)\n",
    "    g.set_yticks([])\n",
    "    g.set_ylabel('')\n",
    "    g.set_xlim((0,28))\n",
    "\n",
    "sel_term = []\n",
    "for j,i in enumerate(x):\n",
    "    sel_ax = cm.fig.add_subplot(gs[j+1, 2])\n",
    "    enr = gp.enrichr(gene_list=[ha.upper() for ha in i],\n",
    "             gene_sets=KEGG,\n",
    "             background = tpm.shape[0],\n",
    "             outdir='test/enrichr_kegg2',\n",
    "             cutoff=0.5,\n",
    "             verbose=False)\n",
    "    temp = enr.res2d[~(\n",
    "        enr.res2d['Term'].isin(exclude_KEGG) |\n",
    "        (enr.res2d['Term'].str.contains('isease')) |\n",
    "         (enr.res2d['Term'].str.contains('arcinoma')) |\n",
    "          (enr.res2d['Term'].str.contains('ancer')) |\n",
    "        (enr.res2d['Term'].str.contains('thyroid')) |\n",
    "        (enr.res2d['Term'].str.contains('yndrome')) |\n",
    "        (enr.res2d['Term'].str.contains('strogen')) |\n",
    "        (enr.res2d['Term'].str.contains('elanoge')) |\n",
    "        (enr.res2d['Term'].str.contains('iabet')) |\n",
    "        (enr.res2d['Term'].str.contains('ysosome')) |\n",
    "        (enr.res2d['Term'].str.contains('nfection')) |\n",
    "        (enr.res2d['Term'].str.contains('homolog')) |\n",
    "        (enr.res2d['Term'].str.contains('Proteasome')) |\n",
    "        (enr.res2d['Term'].str.contains('ntestin')) |\n",
    "        (enr.res2d['Term'].str.contains('cardi')) |\n",
    "        (enr.res2d['Term'].str.contains('Cardi')) |\n",
    "        (enr.res2d['Term'].str.contains('nemia')) |\n",
    "        (enr.res2d['Term'].str.contains('mmunodef'))\n",
    "    )]\n",
    "    temp = temp.sort_values('P-value', ascending = True).iloc[0:5]\n",
    "    temp2 = temp[['Term','Adjusted P-value']]\n",
    "    sel_term.append(temp2['Term'].tolist())\n",
    "\n",
    "    temp2['Term'] = temp2['Term'].str.replace('endoplasmic reticulum', 'ER')\n",
    "    temp2['-Log10(P-Adj)'] = -np.log10(temp2['Adjusted P-value'])\n",
    "    sns.barplot(data = temp2, x = '-Log10(P-Adj)', y = 'Term', ax = sel_ax, color = color_mapping[j])\n",
    "    sel_ax.yaxis.tick_right()\n",
    "    sel_ax.set_yticklabels([i for i in temp['Term']], fontsize = 15)\n",
    "    sel_ax.set_xticks([0])\n",
    "    if j == len(x) - 1:\n",
    "        sel_ax.set_xlabel('-Log10(P-Adj)', fontweight = 'bold', fontsize = 20)\n",
    "        sel_ax.set_xticks([0])\n",
    "        sel_ax.set_xticklabels([], fontweight = 'bold', fontsize = 15)\n",
    "        sns.despine(ax = sel_ax, left = False, bottom = False, right = True, top = True)\n",
    "    else:\n",
    "        sel_ax.set_xlabel('')\n",
    "        sel_ax.set_xticks([])\n",
    "        sns.despine(ax = sel_ax, left = False, bottom = True, right = True, top = True)\n",
    "    sel_ax.set_ylabel('')\n",
    "    sel_ax.tick_params(right=False, bottom=False)\n",
    "    \n",
    "for cl in sorted(clusters['cluster'].unique().astype(int)):\n",
    "    sel_ax = cm.fig.add_subplot(gs[cl+1, 4])\n",
    "    topx = 5\n",
    "    temp = pd.DataFrame()\n",
    "    for j in sel_term[cl]:\n",
    "        gn = [i.capitalize() for i in KEGG[j]]\n",
    "        temp = pd.concat([temp, clusters[clusters['cluster'] == cl].reindex(gn).dropna().sort_values('degree', ascending = False).iloc[0:1].reset_index()])\n",
    "#     temp = clusters[clusters['cluster'] == cl].reindex(sel_term[cl]).dropna().sort_values('degree', ascending = False).iloc[0:topx].reset_index()\n",
    "    sns.scatterplot(data = temp, x = 'degree', y = 'index', ax = sel_ax, color = color_mapping[cl], size = 'norm_deg', legend = False)\n",
    "    sel_ax.yaxis.tick_right()\n",
    "    sel_ax.set_yticklabels([i for i in temp['index']], fontsize = 15)\n",
    "    sel_ax.set_xticks([0])\n",
    "    if cl == clusters['cluster'].max():\n",
    "        sel_ax.set_xlabel('Top Related Nodes\\n(Degree)', fontweight = 'bold', fontsize = 20)\n",
    "        sel_ax.set_xticks([0])\n",
    "        sel_ax.set_xticklabels([], fontweight = 'bold', fontsize = 15)\n",
    "        sns.despine(ax = sel_ax, left = False, bottom = False, right = True, top = True)\n",
    "    else:\n",
    "        sel_ax.set_xlabel('')\n",
    "        sel_ax.set_xticks([])\n",
    "        sns.despine(ax = sel_ax, left = False, bottom = True, right = True, top = True)\n",
    "    sel_ax.set_ylabel('')\n",
    "    sel_ax.tick_params(right=False, bottom=False)\n",
    "\n",
    "plt.savefig('Figures/Figure3C.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d350cace-01be-4838-a800-4eeea104f87e",
   "metadata": {},
   "source": [
    "## Figure 3D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94b7f0c0-a823-4d19-9494-154ea7ff2808",
   "metadata": {},
   "outputs": [],
   "source": [
    "pft1 = ['Stiffness Index (PFT)', 'Peripheral Airway Resistance (PFT)', 'Central Airway Resistance (PFT)']\n",
    "pft2 = ['Forced Expiratory Volume (PFT)', 'Forced Vital Capacity (PFT)', 'Inspiratory Capacity (PFT) ']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e28e612-0c75-4cf0-a1b3-e8ad6aa7b2c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "topx = 0.25\n",
    "final_edges = k.network_ori[k.network_ori['weight']>k.network_ori['weight'].quantile(1-topx)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6aa2f800-1043-4247-bd9d-de0438838f94",
   "metadata": {},
   "outputs": [],
   "source": [
    "net_pft2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "740cbed0-c422-45a3-936e-dafc9ba1d3ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "net_pft1 = final_edges[final_edges['level_0'].isin(pft1) | final_edges['level_1'].isin(pft1)].sort_values('weight', ascending = False)\n",
    "net_pft2 = final_edges[final_edges['level_0'].isin(pft2) | final_edges['level_1'].isin(pft2)].sort_values('weight', ascending = False)\n",
    "net_pft1.to_csv('Networks/GCN/net_pft1.txt', sep = '\\t', index = False)\n",
    "net_pft2.to_csv('Networks/GCN/net_pft2.txt', sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8fc9f78-f798-4489-af11-f82d84c4d34c",
   "metadata": {},
   "source": [
    "## Figure 4A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52b304f0-7d8a-48dd-b4b8-3f813902ac3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = supp1.parse('Metadata and PFTs', index_col = 0)\n",
    "metadata = temp.iloc[0:, 0:2]\n",
    "metadata['Conditions'] = [\"%s (%s)\" % (metadata['Treatment'][i], metadata['Day'][i]) if metadata['Treatment'][i] != 'Control' else 'Control' for i in metadata.index]\n",
    "\n",
    "data = supp3.parse('Putative Metabolomics', index_col = 0)[metadata.index]\n",
    "\n",
    "mapping_putative = supp3.parse('Putative Mapping', index_col = 0)\n",
    "\n",
    "## Remove data that were detected in less than 2 samples per group\n",
    "remove = data[(pd.concat([data.T.notna(),metadata['Conditions']],1).groupby('Conditions').sum() < 2).sum() > 0].index.tolist()\n",
    "removed_mets = mapping_putative.reindex(remove)\n",
    "data_putative = data[~data.index.isin(remove)]\n",
    "\n",
    "Control = impute_knn(data_putative.T.reindex(metadata[metadata['Conditions'] == 'Control'].index).dropna(how = 'all').dropna(how = 'all', axis = 1))\n",
    "Bleo_7 = impute_knn(data_putative.T.reindex(metadata[metadata['Conditions'] == 'Bleomycin (7 Days)'].index).dropna(how = 'all').dropna(how = 'all', axis = 1))\n",
    "Bleo_14 = impute_knn(data_putative.T.reindex(metadata[metadata['Conditions'] == 'Bleomycin (14 Days)'].index).dropna(how = 'all').dropna(how = 'all', axis = 1))\n",
    "Bleo_21 = impute_knn(data_putative.T.reindex(metadata[metadata['Conditions'] == 'Bleomycin (21 Days)'].index).dropna(how = 'all').dropna(how = 'all', axis = 1))\n",
    "Bleo_28 = impute_knn(data_putative.T.reindex(metadata[metadata['Conditions'] == 'Bleomycin (28 Days)'].index).dropna(how = 'all').dropna(how = 'all', axis = 1))\n",
    "\n",
    "\n",
    "data_putative_imputed = pd.concat([Control, Bleo_7, Bleo_14, Bleo_21, Bleo_28]).T.dropna()\n",
    "\n",
    "norm_data_putative_imputed = zscore(data_putative_imputed.T).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e0cad7c-f44d-4bed-b83c-81e71bd34857",
   "metadata": {},
   "outputs": [],
   "source": [
    "var = 'Conditions'\n",
    "sklearn_pca = sklearnPCA(n_components=2)\n",
    "Y_sklearn = sklearn_pca.fit_transform((norm_data_putative_imputed).T)\n",
    "plt.figure(figsize=(6,5))\n",
    "coordinate = -1*pd.DataFrame(Y_sklearn, columns = ['x', 'y'], index = norm_data_putative_imputed.columns) ## -1 is to switch 7 days to left.\n",
    "coordinate = pd.concat([coordinate, metadata],1)\n",
    "ax = sns.scatterplot(data = coordinate, x = 'x', y = 'y', s = 300, hue = var, palette = 'colorblind')\n",
    "ax.set_xlabel('PC1 (%.2f%%)' % (sklearn_pca.explained_variance_ratio_[0]*100), fontsize = 20, fontweight = 'bold')\n",
    "ax.set_ylabel('PC2 (%.2f%%)' % (sklearn_pca.explained_variance_ratio_[1]*100), fontsize = 20, fontweight = 'bold')\n",
    "ax.set_xticklabels([])\n",
    "ax.set_yticklabels([])\n",
    "ax.get_legend().set_title('Groups')\n",
    "plt.setp(ax.get_legend().get_title(), fontsize='15', fontweight = 'bold')\n",
    "plt.setp(ax.get_legend().get_texts(), fontsize='15')\n",
    "plt.tight_layout()\n",
    "plt.legend(bbox_to_anchor=(1.05, 1.0), loc='upper left')\n",
    "plt.savefig('Figures/Figure4A.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c9e969f-3821-4687-b654-ab0f9cc49b10",
   "metadata": {},
   "source": [
    "## Statistical Analysis of Metabolomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff867df0-1390-402f-8d44-d89cd160ca34",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = supp1.parse('Metadata and PFTs', index_col = 0)\n",
    "metadata = temp.iloc[0:, 0:2]\n",
    "metadata['Conditions'] = [\"%s (%s)\" % (metadata['Treatment'][i], metadata['Day'][i]) if metadata['Treatment'][i] != 'Control' else 'Control' for i in metadata.index]\n",
    "\n",
    "data = supp3.parse('Putative Metabolomics', index_col = 0)[metadata.index]\n",
    "\n",
    "mapping_putative = supp3.parse('Putative Mapping', index_col = 0)\n",
    "\n",
    "## Remove data that were detected in less than 2 samples per group\n",
    "remove = data[(pd.concat([data.T.notna(),metadata['Conditions']],1).groupby('Conditions').sum() < 2).sum() > 0].index.tolist()\n",
    "removed_mets = mapping_putative.reindex(remove)\n",
    "data_putative = data[~data.index.isin(remove)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6461db4-f645-49d6-b46d-54b156e12ebc",
   "metadata": {},
   "outputs": [],
   "source": [
    "Control = impute_knn(data_putative.T.reindex(metadata[metadata['Conditions'] == 'Control'].index).dropna(how = 'all').dropna(how = 'all', axis = 1))\n",
    "Bleo_7 = impute_knn(data_putative.T.reindex(metadata[metadata['Conditions'] == 'Bleomycin (7 Days)'].index).dropna(how = 'all').dropna(how = 'all', axis = 1))\n",
    "Bleo_14 = impute_knn(data_putative.T.reindex(metadata[metadata['Conditions'] == 'Bleomycin (14 Days)'].index).dropna(how = 'all').dropna(how = 'all', axis = 1))\n",
    "Bleo_21 = impute_knn(data_putative.T.reindex(metadata[metadata['Conditions'] == 'Bleomycin (21 Days)'].index).dropna(how = 'all').dropna(how = 'all', axis = 1))\n",
    "Bleo_28 = impute_knn(data_putative.T.reindex(metadata[metadata['Conditions'] == 'Bleomycin (28 Days)'].index).dropna(how = 'all').dropna(how = 'all', axis = 1))\n",
    "\n",
    "res = []\n",
    "\n",
    "for i in data_putative.index:\n",
    "    temp1 = ttest_ind(Bleo_7[i].dropna(), Control[i].dropna(), nan_policy='omit')\n",
    "    temp2 = ttest_ind(Bleo_14[i].dropna(), Control[i].dropna(), nan_policy='omit')\n",
    "    temp3 = ttest_ind(Bleo_21[i].dropna(), Control[i].dropna(), nan_policy='omit')\n",
    "    temp4 = ttest_ind(Bleo_28[i].dropna(), Control[i].dropna(), nan_policy='omit')\n",
    "    res.append([\n",
    "                i, \n",
    "                np.log(Bleo_7[i].mean()/Control[i].mean()), temp1[1],\n",
    "                np.log(Bleo_14[i].mean()/Control[i].mean()), temp2[1],\n",
    "                np.log(Bleo_21[i].mean()/Control[i].mean()), temp3[1],\n",
    "                np.log(Bleo_28[i].mean()/Control[i].mean()), temp4[1]\n",
    "               ])\n",
    "res = pd.DataFrame(res,columns=['Metabolite',\n",
    "                                'Log2FoldChange (Bleo (7 Days) vs Control)', 'P-value (Bleo (7 Days) vs Control)',\n",
    "                                'Log2FoldChange (Bleo (14 Days) vs Control)', 'P-value (Bleo (14 Days) vs Control)',\n",
    "                                'Log2FoldChange (Bleo (21 Days) vs Control)', 'P-value (Bleo (21 Days) vs Control)',\n",
    "                                'Log2FoldChange (Bleo (28 Days) vs Control)', 'P-value (Bleo (28 Days) vs Control)'\n",
    "                                ]).set_index('Metabolite')\n",
    "\n",
    "res['P-Adj (Bleo (7 Days) vs Control)'] = multipletests(res['P-value (Bleo (7 Days) vs Control)'].fillna(1),method='fdr_bh')[1]\n",
    "res['P-Adj (Bleo (14 Days) vs Control)'] = multipletests(res['P-value (Bleo (14 Days) vs Control)'].fillna(1),method='fdr_bh')[1]\n",
    "res['P-Adj (Bleo (21 Days) vs Control)'] = multipletests(res['P-value (Bleo (21 Days) vs Control)'].fillna(1),method='fdr_bh')[1]\n",
    "res['P-Adj (Bleo (28 Days) vs Control)'] = multipletests(res['P-value (Bleo (28 Days) vs Control)'].fillna(1),method='fdr_bh')[1]\n",
    "res=pd.concat([mapping_putative.reindex(res.index),res],1)\n",
    "\n",
    "res['DIRECTION (Bleo (7 Days) vs Control)'] = ['UP' if i > 0 else 'DOWN' for i in res['Log2FoldChange (Bleo (7 Days) vs Control)']]\n",
    "res['DIRECTION (Bleo (14 Days) vs Control)'] = ['UP' if i > 0 else 'DOWN' for i in res['Log2FoldChange (Bleo (14 Days) vs Control)']]\n",
    "res['DIRECTION (Bleo (21 Days) vs Control)'] = ['UP' if i > 0 else 'DOWN' for i in res['Log2FoldChange (Bleo (21 Days) vs Control)']]\n",
    "res['DIRECTION (Bleo (28 Days) vs Control)'] = ['UP' if i > 0 else 'DOWN' for i in res['Log2FoldChange (Bleo (28 Days) vs Control)']]\n",
    "\n",
    "res = res[mapping_putative.columns.tolist()+ [\n",
    "    'Log2FoldChange (Bleo (7 Days) vs Control)', 'P-value (Bleo (7 Days) vs Control)','P-Adj (Bleo (7 Days) vs Control)', 'DIRECTION (Bleo (7 Days) vs Control)',\n",
    "    'Log2FoldChange (Bleo (14 Days) vs Control)', 'P-value (Bleo (14 Days) vs Control)','P-Adj (Bleo (14 Days) vs Control)', 'DIRECTION (Bleo (14 Days) vs Control)',\n",
    "    'Log2FoldChange (Bleo (21 Days) vs Control)', 'P-value (Bleo (21 Days) vs Control)','P-Adj (Bleo (21 Days) vs Control)','DIRECTION (Bleo (21 Days) vs Control)',\n",
    "    'Log2FoldChange (Bleo (28 Days) vs Control)', 'P-value (Bleo (28 Days) vs Control)','P-Adj (Bleo (28 Days) vs Control)','DIRECTION (Bleo (28 Days) vs Control)'\n",
    "]]\n",
    "\n",
    "res.to_csv('StatisticalAnalysis/Metabolomics/Results.txt', sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23717cb0-ad84-43bd-b1d3-f7b8756822e6",
   "metadata": {},
   "source": [
    "## Figure 4B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7020d1a0-b851-4dcf-8d5c-ba66fc0cd2ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = pd.read_csv('StatisticalAnalysis/Metabolomics/Results.txt', sep = '\\t', index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e794364a-bc11-47f2-abdf-7d63f6bcde4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_padj = res[['P-Adj (Bleo (7 Days) vs Control)', 'P-Adj (Bleo (14 Days) vs Control)', 'P-Adj (Bleo (21 Days) vs Control)', 'P-Adj (Bleo (28 Days) vs Control)']]\n",
    "res_padj.columns = res_padj.columns.str.replace('P-Adj', '')\n",
    "res = res[['DIRECTION (Bleo (7 Days) vs Control)', 'DIRECTION (Bleo (14 Days) vs Control)', 'DIRECTION (Bleo (21 Days) vs Control)', 'DIRECTION (Bleo (28 Days) vs Control)']]\n",
    "res.columns = res.columns.str.replace('DIRECTION', '')\n",
    "res = res[res_padj < 0.05]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7730a331-21e3-41d6-9f4e-11619fd6a733",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = res.unstack().reset_index().dropna().groupby(['level_0',0]).count().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f453d0ab-97b3-4a61-8efc-55d3e05db7cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize = (5,5))\n",
    "g = sns.barplot(data = temp, x = 'level_0', y = 'Metabolite', hue = 0, order = res.columns, hue_order = ['UP', 'DOWN'])\n",
    "g.set_yticklabels([int(i) for i in g.get_yticks()], fontsize = 15)\n",
    "g.set_xticklabels(g.get_xticklabels(),rotation=30, ha = 'right', fontsize = 15, fontweight = 'bold')\n",
    "g.set_xlabel('')\n",
    "g.set_ylabel('# of DM', fontsize= 20, fontweight = 'bold')\n",
    "g.get_legend().set_title('Direction')\n",
    "plt.setp(g.get_legend().get_title(), fontsize='15', fontweight = 'bold')\n",
    "plt.setp(g.get_legend().get_texts(), fontsize='15')\n",
    "plt.tight_layout()\n",
    "plt.savefig('Figures/Figure4B.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90b38e77-bbc8-4fb4-a270-99fd95cffb61",
   "metadata": {},
   "source": [
    "## Figure 4C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1ba332b-818a-402b-bfed-10c6c092c1aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = pd.read_csv('StatisticalAnalysis/Metabolomics/Results.txt', sep = '\\t', index_col = 0)\n",
    "res = res[['P-Adj (Bleo (7 Days) vs Control)', 'P-Adj (Bleo (14 Days) vs Control)', 'P-Adj (Bleo (21 Days) vs Control)', 'P-Adj (Bleo (28 Days) vs Control)']]\n",
    "res = res[res < 0.05].dropna(thresh = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7945df92-6102-45d1-907b-bf07718e9155",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = supp1.parse('Metadata and PFTs', index_col = 0)\n",
    "metadata = temp.iloc[0:, 0:2]\n",
    "metadata['Conditions'] = [\"%s (%s)\" % (metadata['Treatment'][i], metadata['Day'][i]) if metadata['Treatment'][i] != 'Control' else 'Control' for i in metadata.index]\n",
    "\n",
    "data = supp3.parse('Putative Metabolomics', index_col = 0)[metadata.index]\n",
    "\n",
    "mapping_putative = supp3.parse('Putative Mapping', index_col = 0)\n",
    "\n",
    "## Remove data that were detected in less than 2 samples per group\n",
    "remove = data[(pd.concat([data.T.notna(),metadata['Conditions']],1).groupby('Conditions').sum() < 2).sum() > 0].index.tolist()\n",
    "removed_mets = mapping_putative.reindex(remove)\n",
    "data_putative = data[~data.index.isin(remove)]\n",
    "\n",
    "Control = impute_knn(data_putative.T.reindex(metadata[metadata['Conditions'] == 'Control'].index).dropna(how = 'all').dropna(how = 'all', axis = 1))\n",
    "Bleo_7 = impute_knn(data_putative.T.reindex(metadata[metadata['Conditions'] == 'Bleomycin (7 Days)'].index).dropna(how = 'all').dropna(how = 'all', axis = 1))\n",
    "Bleo_14 = impute_knn(data_putative.T.reindex(metadata[metadata['Conditions'] == 'Bleomycin (14 Days)'].index).dropna(how = 'all').dropna(how = 'all', axis = 1))\n",
    "Bleo_21 = impute_knn(data_putative.T.reindex(metadata[metadata['Conditions'] == 'Bleomycin (21 Days)'].index).dropna(how = 'all').dropna(how = 'all', axis = 1))\n",
    "Bleo_28 = impute_knn(data_putative.T.reindex(metadata[metadata['Conditions'] == 'Bleomycin (28 Days)'].index).dropna(how = 'all').dropna(how = 'all', axis = 1))\n",
    "\n",
    "\n",
    "data_putative_imputed = pd.concat([Control, Bleo_7, Bleo_14, Bleo_21, Bleo_28]).T.dropna()\n",
    "\n",
    "norm_data_putative_imputed = zscore(data_putative_imputed.T).T.reindex(res.index).rename(columns = metadata['Conditions'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4dad5ae5-efc0-4ca3-a137-9ea66999020a",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_linkage = linkage(\n",
    "    distance.pdist(norm_data_putative_imputed.fillna(0)), method='average')\n",
    "clust = pd.Series([i[0] for i in cut_tree(row_linkage, n_clusters = 2)], \n",
    "                  index = norm_data_putative_imputed.index).sort_values().replace({0:1, 1:0}) ## Just to put the up-regulated clusters as DM-0\n",
    "\n",
    "clust.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce37bd01-e859-43f3-ad15-4b0b70dcfb99",
   "metadata": {},
   "outputs": [],
   "source": [
    "vmin = norm_data_putative_imputed.min().min()\n",
    "vmax = norm_data_putative_imputed.max().max()\n",
    "order = ['Control', 'Bleomycin (7 Days)', 'Bleomycin (14 Days)', 'Bleomycin (21 Days)', 'Bleomycin (28 Days)']\n",
    "cmap=matplotlib.colors.LinearSegmentedColormap.from_list(\"\", [\"#0000a5\",'#0000d8',\"#FFFAF0\",'#d80000',\"#a50000\"])\n",
    "g = sns.clustermap(norm_data_putative_imputed.reindex(clust.index[::-1]).fillna(0)[order], figsize = (5,8), xticklabels = 1, yticklabels = 0, center = 0, cmap = 'RdYlBu_r', col_cluster = False, row_cluster = False, row_linkage = row_linkage, vmin = vmin, vmax = vmax)\n",
    "g.ax_heatmap.vlines(metadata['Conditions'].value_counts()[order].cumsum(), ymin = g.ax_heatmap.get_ylim()[1], ymax = g.ax_heatmap.get_ylim()[0], color = 'k')\n",
    "g.ax_heatmap.hlines(clust.value_counts().sort_index(ascending = False)[::-1].cumsum().tolist(), xmin = g.ax_heatmap.get_xlim()[1], xmax = g.ax_heatmap.get_xlim()[0], color = 'k')\n",
    "g.ax_heatmap.tick_params(right=False, bottom=False)\n",
    "xticklabels = metadata['Conditions'].value_counts().sort_index()[order]\n",
    "g.ax_heatmap.set_xticks(xticklabels.cumsum()-((xticklabels/2)))\n",
    "g.ax_heatmap.set_xticklabels(xticklabels.index.tolist(), fontsize = 15, fontweight = 'bold', rotation = 30, ha = 'right')\n",
    "g.savefig('Figures/Figure4C.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3264190-fa3c-4b73-b0ea-5162f6143118",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(figsize = (3, 8), nrows = 2, ncols = 1)\n",
    "color_mapping = dict(zip(clust.value_counts().index,sns.color_palette(\"tab10\").as_hex()))\n",
    "\n",
    "for i, cl in enumerate(clust.value_counts().index):\n",
    "    days = {'A': 0, 'B': 7, 'C': 14, 'D': 21, 'E': 28}\n",
    "    x_interp = np.linspace(days['A'], days['E'], 100)\n",
    "    sel_col = ['Control', 'Bleomycin (7 Days)', 'Bleomycin (14 Days)', 'Bleomycin (21 Days)', 'Bleomycin (28 Days)']\n",
    "    mapping = dict(zip(sel_col, ['A', 'B', 'C', 'D', 'E']))\n",
    "    sel_data = norm_data_putative_imputed[sel_col].reindex(clust[clust == cl].index).rename(columns = mapping).T.reset_index().groupby('index').mean().T[['A', 'B', 'C', 'D', 'E']].rename(columns = days).dropna()\n",
    "    y_interp = interp1d(list(days.values()), list(sel_data.values), kind = 'quadratic')(x_interp)\n",
    "    data_interp = pd.DataFrame(y_interp, columns = x_interp).unstack().reset_index().groupby('level_0').mean().reset_index()\n",
    "    data_interp[0] = data_interp[0] - data_interp.loc[0,0]\n",
    "    temp_max = data_interp.copy()[0]\n",
    "    g = sns.lineplot(data = data_interp, x = 'level_0', y = 0, ax = axs[i], color = color_mapping[cl], linewidth=5, zorder = 15)\n",
    "    \n",
    "    if cl == clust.value_counts().index.tolist()[-1]:\n",
    "        g.set_xlabel('Days', fontweight = 'bold', fontsize = 20)\n",
    "        g.set_xticks(list(days.values()))\n",
    "        g.set_xticklabels(list(days.values()), fontweight = 'bold', fontsize = 15)\n",
    "    else:\n",
    "        g.set_xlabel('')\n",
    "        g.set_xticks([])\n",
    "    temp_ylim = tuple(g.get_ylim())\n",
    "    g.vlines([7, 14, 21], ymin = g.get_ylim()[0], ymax = g.get_ylim()[1], color = 'lightgray', zorder = 10)\n",
    "    g.set_ylim(temp_ylim)\n",
    "    \n",
    "    g.set_yticks([])\n",
    "    g.set_ylabel('')\n",
    "    g.set_xlim((0,28))\n",
    "    \n",
    "plt.savefig('Figures/Figure4C_trend.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2bfa40a-f951-447b-9fac-00e3bf9ddc68",
   "metadata": {},
   "source": [
    "## Figure 4D-E\n",
    "\n",
    "Do figure 5C first, because the metabolite enrichment will use the DM-0 and DM-1 assignment calculated there"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b7bb2ee-52c8-4057-a204-7c9ab31b473a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pathways = supp3.parse('Pathway Mapping').dropna()\n",
    "pathways = pathways[pathways['ID'] != '-']\n",
    "pathways = pathways.groupby('Pathway')['ID'].apply(lambda x: list(x.unique())).to_dict()\n",
    "pw_bg = mapping_putative.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d8dfffc-7fe7-477b-9790-ec3d80735e71",
   "metadata": {},
   "outputs": [],
   "source": [
    "cl = 1\n",
    "temp = mapping_putative.reindex(clust[clust == cl].index)['HMDB ID'].dropna()\n",
    "met_list = []\n",
    "for i in temp.tolist():\n",
    "    if i.count('HMDB') == 0:\n",
    "        pass\n",
    "    elif i.count('HMDB') > 1:\n",
    "        for j in i.replace(',','\\n').split('\\n'):\n",
    "            met_list.append(j)\n",
    "    else:\n",
    "        met_list.append(i.replace('\\n',''))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c135946-1655-44c9-944c-2fc886553255",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(figsize = (3, 8), nrows = 2, ncols = 1, sharex = False)\n",
    "color_mapping = dict(zip(clust.value_counts().index,sns.color_palette(\"tab10\").as_hex()))\n",
    "\n",
    "for cl in clust.unique()[::-1]:\n",
    "    print(cl)\n",
    "    met_list = clust[clust == cl].index.tolist()\n",
    "    x = []\n",
    "    for pw in list(pathways.keys()):\n",
    "        hist_mets = mapping_putative.reindex(clust.reindex(list(set(met_list).intersection(set(pathways[pw])))).index)['Compound name']\n",
    "        hits = len(hist_mets)\n",
    "        num_ofpw = len(set(sum(list(pathways.values()), [])))\n",
    "        num_ofmets_inpw = len(set(pathways[pw]))\n",
    "        total_mets = len(set(met_list))\n",
    "        pval = 1-hypergeom.cdf(hits,num_ofpw-num_ofmets_inpw,total_mets,num_ofmets_inpw)\n",
    "        x.append([pw, hits, num_ofmets_inpw, ';'.join(list(hist_mets)), pval])\n",
    "    x = pd.DataFrame(x, columns = ['Pathways', '# of Hits', 'Total Mets', 'Hits Metabolites', 'pval']).set_index('Pathways')\n",
    "    x = x[(x['# of Hits'] > 0) & (x['Total Mets'] > 1)]\n",
    "    x = x[~x.index.isin(['Prostate cancer'])]\n",
    "    print(x[x['pval'] < 0.05].shape[0])\n",
    "    temp = x[x['pval'] < 0.05].sort_values('pval').iloc[0:10].reset_index()\n",
    "    temp['-Log10(P)'] = -np.log10(temp['pval'] + 1E-2)\n",
    "    temp['color'] = '#a50000'\n",
    "    g = sns.barplot(data = temp, x = '-Log10(P)', y = 'Pathways', color = color_mapping[cl], ax = axs[cl])\n",
    "    \n",
    "    g.set_ylabel('')\n",
    "    sns.despine()\n",
    "    g.set_yticklabels(g.get_yticklabels(), fontsize = 15)\n",
    "    if cl == 0:\n",
    "        g.set_xlabel('')\n",
    "    else:\n",
    "        g.set_xlabel('-Log10(P)', fontsize = 20, fontweight = 'bold')\n",
    "    g.set_xticklabels(['%.1f' % i for i in g.get_xticks()], fontsize = 15)\n",
    "    g.set_title('DM-%d' % cl, fontsize = 20, fontweight = 'bold')\n",
    "plt.savefig('Figures/Figure4D-E.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f96da44c-613b-4565-b0e3-3507102b5e52",
   "metadata": {},
   "source": [
    "### Loading Pre-analyzed networks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd842783-eb42-42b0-89e6-d13cf1c4e67f",
   "metadata": {},
   "outputs": [],
   "source": [
    "k = pickle.load(open('Networks/ReadyObjects/MCN.pkl', 'rb'))\n",
    "\n",
    "# iGraph object is under k.network\n",
    "# Pandas object of the original network (not filtered based on the correlation threshold) is under k.network_ori\n",
    "# Subnetwork info iss under k.clustering\n",
    "# Degree centrality is under k.degree"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe548387-745d-4a10-a51d-60d640d1d05a",
   "metadata": {},
   "source": [
    "## Figure 5A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "276bfbe0-906f-48f6-97b8-13f5f522eec5",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = []\n",
    "for cl in sorted(k.clustering.unique()):\n",
    "    k_temp = k.network.subgraph(k.clustering[k.clustering == cl].index)\n",
    "    print(cl)\n",
    "    print(k_temp.transitivity_avglocal_undirected())\n",
    "    x.append([cl, k_temp.transitivity_avglocal_undirected(), cl])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "871a29d9-1bc6-4c76-9cfa-1e29aad4fb86",
   "metadata": {},
   "source": [
    "## Figure 5B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28ad7e26-bc57-41a3-9668-d3519b52ce69",
   "metadata": {},
   "outputs": [],
   "source": [
    "k = pickle.load(open('Networks/ReadyObjects/MCN.pkl', 'rb'))\n",
    "mapping_putative = supp3.parse('Putative Mapping', index_col = 0)\n",
    "\n",
    "clusters = pd.concat([k.clustering.rename('cluster'), k.degree],1)\n",
    "clusters = clusters[clusters['cluster'].notna()]\n",
    "clusters['norm_deg'] = ((clusters['degree'] - clusters['degree'].min())/(clusters['degree'].max() - clusters['degree'].min()))\n",
    "clusters = pd.concat([clusters, mapping_putative.reindex(clusters.index)], 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "948ecb92-6f02-4a03-ab51-13b463c03766",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = supp1.parse('Metadata and PFTs', index_col = 0)\n",
    "metadata = temp.iloc[0:, 0:2]\n",
    "metadata['Conditions'] = [\"%s (%s)\" % (metadata['Treatment'][i], metadata['Day'][i]) if metadata['Treatment'][i] != 'Control' else 'Control' for i in metadata.index]\n",
    "PFT = temp.iloc[0:, 2:]\n",
    "norm_pft = zscore(PFT).T.rename(columns = metadata['Conditions'].to_dict())\n",
    "\n",
    "data = supp3.parse('Putative Metabolomics', index_col = 0)[metadata.index]\n",
    "\n",
    "mapping_putative = supp3.parse('Putative Mapping', index_col = 0)\n",
    "\n",
    "## Remove data that were detected in less than 2 samples per group\n",
    "remove = data[(pd.concat([data.T.notna(),metadata['Conditions']],1).groupby('Conditions').sum() < 2).sum() > 0].index.tolist()\n",
    "removed_mets = mapping_putative.reindex(remove)\n",
    "data_putative = data[~data.index.isin(remove)]\n",
    "\n",
    "Control = impute_knn(data_putative.T.reindex(metadata[metadata['Conditions'] == 'Control'].index).dropna(how = 'all').dropna(how = 'all', axis = 1))\n",
    "Bleo_7 = impute_knn(data_putative.T.reindex(metadata[metadata['Conditions'] == 'Bleomycin (7 Days)'].index).dropna(how = 'all').dropna(how = 'all', axis = 1))\n",
    "Bleo_14 = impute_knn(data_putative.T.reindex(metadata[metadata['Conditions'] == 'Bleomycin (14 Days)'].index).dropna(how = 'all').dropna(how = 'all', axis = 1))\n",
    "Bleo_21 = impute_knn(data_putative.T.reindex(metadata[metadata['Conditions'] == 'Bleomycin (21 Days)'].index).dropna(how = 'all').dropna(how = 'all', axis = 1))\n",
    "Bleo_28 = impute_knn(data_putative.T.reindex(metadata[metadata['Conditions'] == 'Bleomycin (28 Days)'].index).dropna(how = 'all').dropna(how = 'all', axis = 1))\n",
    "\n",
    "\n",
    "data_putative_imputed = pd.concat([Control, Bleo_7, Bleo_14, Bleo_21, Bleo_28]).T.dropna()\n",
    "\n",
    "norm_data_putative_imputed = zscore(data_putative_imputed.T).T.rename(columns = metadata['Conditions'].to_dict())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a117d46-e6ad-4283-a2f2-f37967c8131c",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = []\n",
    "for cl in sorted(clusters['cluster'].unique()):\n",
    "    sel_data = norm_data_putative_imputed.reindex(clusters[clusters['cluster'] == cl].index).rename(columns = metadata['Conditions'].to_dict())#.T#.reset_index().groupby('index').mean().T.rename(columns = days)\n",
    "    row_linkage = linkage(distance.pdist(sel_data.fillna(0)), method='average')\n",
    "    dn = dendrogram(row_linkage, no_plot = True)['leaves']\n",
    "    temp = sel_data.iloc[dn].index.tolist()\n",
    "    x.append(temp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f7c4ad5-94be-4e58-b5fc-a1fd3928ce44",
   "metadata": {},
   "outputs": [],
   "source": [
    "days = {'Control': 0, 'Bleomycin (7 Days)': 7, 'Bleomycin (14 Days)': 14, 'Bleomycin (21 Days)': 21, 'Bleomycin (28 Days)': 28}\n",
    "x_interp = np.linspace(days['Control'], days['Bleomycin (28 Days)'], 100)\n",
    "color_mapping = dict(zip(sorted(clusters['cluster'].unique()),sns.color_palette(\"tab10\").as_hex()[::-1]))\n",
    "\n",
    "cluster_color = pd.Series(clusters['cluster']).map(color_mapping)\n",
    "sel_data = norm_data_putative_imputed.reindex(sum(x,[]))\n",
    "cm = sns.clustermap(sel_data.fillna(0), figsize=(12.63, 10),  yticklabels = 0, xticklabels = 0, center = 0, cmap = 'RdYlBu_r', row_colors = cluster_color, row_cluster = False, col_cluster = False)\n",
    "cm.cax.set_visible(False)\n",
    "cm.gs.update(left=0.05, right=0.25)\n",
    "cm.ax_heatmap.vlines(metadata.groupby('Conditions').count()['Treatment'].cumsum().tolist(), ymin = 0, ymax = sel_data.shape[0], color = 'gray')\n",
    "cm.ax_heatmap.hlines(np.cumsum([len(i) for i in x]), xmin = 0, xmax = sel_data.shape[1], color = 'black')\n",
    "cm.ax_heatmap.set_xlabel('Days', fontweight = 'bold', fontsize = 20)\n",
    "cm.ax_heatmap.set_ylabel('', fontweight = 'bold', fontsize = 20)\n",
    "cm.ax_heatmap.set_xticks((metadata.groupby('Conditions').count().cumsum() - metadata.groupby('Conditions').count()/2)['Treatment'].tolist())\n",
    "cm.ax_heatmap.set_xticklabels([0, 7, 14, 21, 28], fontweight = 'bold', fontsize = 15)\n",
    "\n",
    "gs = GridSpec(int(clusters['cluster'].max()+2),5, top = .95, left = 0.26, hspace = 0.2, \n",
    "              wspace = 0.21,width_ratios=[.5,0.1, .5,  1, 0.25])\n",
    "\n",
    "\n",
    "for cl in sorted(clusters['cluster'].unique().astype(int)):\n",
    "    sel_ax = cm.fig.add_subplot(gs[cl+1, 0])\n",
    "    sel_data = norm_data_putative_imputed.reindex(clusters[clusters['cluster'] == cl].index).rename(columns = metadata['Conditions'].to_dict()).T.reset_index().groupby('index').mean().T.rename(columns = days).dropna()[[0, 7, 14, 21, 28]]\n",
    "    y_interp = interp1d(list(days.values()), list(sel_data.values), kind = 'quadratic')(x_interp)\n",
    "    data_interp = pd.DataFrame(y_interp, columns = x_interp).unstack().reset_index().groupby('level_0').mean().reset_index()\n",
    "    temp_max = data_interp.copy()[0]\n",
    "    g = sns.lineplot(data = data_interp, x = 'level_0', y = 0, ax = sel_ax, color = color_mapping[cl])\n",
    "    sel_data = norm_pft.reindex(clusters[clusters['cluster'] == cl].index).rename(columns = metadata['Conditions'].to_dict()).T.reset_index().groupby('index').mean().T.rename(columns = days).dropna()[[0, 7, 14, 21, 28]]\n",
    "    \n",
    "    if sel_data.shape[0] > 0:\n",
    "        y_interp = interp1d(list(days.values()), list(sel_data.values), kind = 'quadratic')(x_interp)\n",
    "        temp_max = pd.concat([temp_max, data_interp.copy()[0]])\n",
    "        data_interp = pd.DataFrame(y_interp, columns = x_interp).unstack().reset_index().groupby('level_0').mean().reset_index()\n",
    "        g = sns.lineplot(data = data_interp, x = 'level_0', y = 0, ax = sel_ax, color = color_mapping[cl], linestyle='--')\n",
    "    if cl == clusters['cluster'].max():\n",
    "        g.set_xlabel('Days', fontweight = 'bold', fontsize = 20)\n",
    "        g.set_xticks(list(days.values()))\n",
    "        g.set_xticklabels(list(days.values()), fontweight = 'bold', fontsize = 15)\n",
    "    else:\n",
    "        g.set_xlabel('')\n",
    "        g.set_xticks([])\n",
    "    temp_ylim = tuple(g.get_ylim())\n",
    "    g.vlines([7, 14, 21], ymin = g.get_ylim()[0], ymax = g.get_ylim()[1], color = 'lightgray')\n",
    "    g.set_ylim(temp_ylim)\n",
    "    g.set_yticks([])\n",
    "    g.set_ylabel('')\n",
    "    g.set_xlim((0,28))\n",
    "    \n",
    "for cl in sorted(clusters['cluster'].unique().astype(int)):\n",
    "    sel_ax = cm.fig.add_subplot(gs[cl+1, 1])\n",
    "    topx = 5\n",
    "    temp = clusters[clusters['cluster'] == cl].sort_values('degree', ascending = False).iloc[0:topx].reset_index()\n",
    "    temp['index'] = temp['Compound name']\n",
    "    sns.scatterplot(data = temp, x = 'degree', y = 'index', ax = sel_ax, color = color_mapping[cl], size = 'norm_deg', legend = False)\n",
    "    sel_ax.yaxis.tick_right()\n",
    "    sel_ax.set_yticklabels([i for i in temp['index']], fontsize = 15)\n",
    "    sel_ax.set_xticks([0])\n",
    "    if cl == clusters['cluster'].max():\n",
    "        sel_ax.set_xlabel('Top Nodes\\n(Degree)', fontweight = 'bold', fontsize = 20)\n",
    "        sel_ax.set_xticks([0])\n",
    "        sel_ax.set_xticklabels([], fontweight = 'bold', fontsize = 15)\n",
    "        sns.despine(ax = sel_ax, left = False, bottom = False, right = True, top = True)\n",
    "    else:\n",
    "        sel_ax.set_xlabel('')\n",
    "        sel_ax.set_xticks([])\n",
    "        sns.despine(ax = sel_ax, left = False, bottom = True, right = True, top = True)\n",
    "    sel_ax.set_ylabel('')\n",
    "    sel_ax.tick_params(right=False, bottom=False)\n",
    "\n",
    "sel_term = [] \n",
    "for cl in sorted(clusters['cluster'].unique().astype(int)):\n",
    "    sel_ax = cm.fig.add_subplot(gs[cl+1, 3])\n",
    "    print(cl)\n",
    "    met_list = clusters[(clusters['cluster'] == cl)].index.tolist()\n",
    "    y = []\n",
    "    for pw in list(pathways.keys()):\n",
    "        hist_mets = mapping_putative.reindex(clusters['cluster'].reindex(list(set(met_list).intersection(set(pathways[pw])))).index)['Compound name']\n",
    "        hits = len(hist_mets)\n",
    "        num_ofpw = len(set(sum(list(pathways.values()), [])))\n",
    "        num_ofmets_inpw = len(set(pathways[pw]))\n",
    "        total_mets = len(set(met_list))\n",
    "        pval = 1-hypergeom.cdf(hits,num_ofpw-num_ofmets_inpw,total_mets,num_ofmets_inpw)\n",
    "        y.append([pw, hits, num_ofmets_inpw, ';'.join(list(hist_mets)), pval])\n",
    "    y = pd.DataFrame(y, columns = ['Pathways', '# of Hits', 'Total Mets', 'Hits Metabolites', 'pval']).set_index('Pathways')\n",
    "    y = y[(y['# of Hits'] > 0) & (y['Total Mets'] > 1)]\n",
    "    y = y[~y.index.isin(['Prostate cancer', 'Amyotrophic lateral sclerosis (ALS)', 'Photosynthesis'])]\n",
    "\n",
    "    temp = y.sort_values('pval').iloc[0:5].reset_index()\n",
    "    temp['-Log10(P)'] = -np.log10(temp['pval'] + 1E-2)\n",
    "    temp['color'] = '#a50000'\n",
    "    sel_term.append(temp['Pathways'].tolist())\n",
    "    sns.barplot(data = temp, x = '-Log10(P)', y = 'Pathways', color = color_mapping[cl], ax = sel_ax)\n",
    "    \n",
    "    sel_ax.yaxis.tick_right()\n",
    "    sel_ax.set_yticklabels([i for i in temp['Pathways']], fontsize = 15)\n",
    "    sel_ax.set_xticks([0])\n",
    "    if cl == max(clusters['cluster'].unique()):\n",
    "        sel_ax.set_xlabel('-Log10(P-Adj)', fontweight = 'bold', fontsize = 20)\n",
    "        sel_ax.set_xticks([0])\n",
    "        sel_ax.set_xticklabels([], fontweight = 'bold', fontsize = 15)\n",
    "        sns.despine(ax = sel_ax, left = False, bottom = False, right = True, top = True)\n",
    "    else:\n",
    "        sel_ax.set_xlabel('')\n",
    "        sel_ax.set_xticks([])\n",
    "        sns.despine(ax = sel_ax, left = False, bottom = True, right = True, top = True)\n",
    "    sel_ax.set_ylabel('')\n",
    "    sel_ax.tick_params(right=False, bottom=False)\n",
    "\n",
    "plt.savefig('Figures/Figure5B.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e428cd6b-9dcf-434a-95f8-6df99b3aa56e",
   "metadata": {},
   "source": [
    "## Figure 5D-E"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b297f1dd-5c4d-4c04-8a04-1e520fc6afe6",
   "metadata": {},
   "source": [
    "### Integrating GCN and MCN\n",
    "\n",
    "Using the enriched subsystems from Mouse Metabolic Model, retrieved from https://metabolicatlas.org/ on September 9, 2022"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92687e8c-865c-4abd-9399-07a3bcf43427",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = supp1.parse('Metadata and PFTs', index_col = 0)\n",
    "metadata = temp.iloc[0:, 0:2]\n",
    "metadata['Conditions'] = [\"%s (%s)\" % (metadata['Treatment'][i], metadata['Day'][i]) if metadata['Treatment'][i] != 'Control' else 'Control' for i in metadata.index]\n",
    "tpm = supp1.parse('TPM', index_col = 0)[metadata.index]\n",
    "mapping_putative = supp3.parse('Putative Mapping', index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "374e2cfe-3c0e-439a-83b5-ffa21e8ac637",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat = loadmat('lib/Mouse-GEM.mat')['mouseGEM']\n",
    "\n",
    "stoc_mat = pd.DataFrame.sparse.from_spmatrix(mat['S'][0][0])\n",
    "fin = []\n",
    "for i in range(len(mat['subSystems'][0][0])):\n",
    "    if (len(mat['grRules'][0][0][i][0]) > 0) & (stoc_mat[i].sum() != 0):\n",
    "        genes = mat['grRules'][0][0][i][0][0].replace('(', '').replace(')', '').replace('and', 'or').split(' or ')\n",
    "        sub = mat['subSystems'][0][0][i][0][0][0][0]\n",
    "        met = []\n",
    "        for j in stoc_mat[i][stoc_mat[i] != 0].index:\n",
    "            if len(mat['metHMDBID'][0][0][j][0]) > 0:\n",
    "                met.append(mat['metHMDBID'][0][0][j][0][0])\n",
    "            else:\n",
    "                pass\n",
    "        met = list(set(met))\n",
    "        fin.append([sub, genes, met])\n",
    "GMT_genes = pd.DataFrame(fin, columns = ['subsystem', 'genes', 'hmdb']).groupby('subsystem')['genes'].apply(lambda x: list(set(sum(x,[])))).to_dict()\n",
    "GMT_mets = pd.DataFrame(fin, columns = ['subsystem', 'genes', 'hmdb']).groupby('subsystem')['hmdb'].apply(lambda x: list(set(sum(x,[])))).to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3262b20-bdaf-4c0b-916e-ca299858d38b",
   "metadata": {},
   "outputs": [],
   "source": [
    "k = pickle.load(open('Networks/ReadyObjects/GCN.pkl', 'rb'))\n",
    "\n",
    "clusters_GCN = pd.concat([k.clustering.rename('cluster'), k.degree],1)\n",
    "clusters_GCN = clusters_GCN[clusters_GCN['cluster'].notna()]\n",
    "\n",
    "k = pickle.load(open('Networks/ReadyObjects/MCN.pkl', 'rb'))\n",
    "mapping_putative = supp3.parse('Putative Mapping', index_col = 0)\n",
    "\n",
    "clusters_MCN = pd.concat([k.clustering.rename('cluster'), k.degree],1)\n",
    "clusters_MCN = clusters_MCN[clusters_MCN['cluster'].notna()]\n",
    "clusters_MCN = pd.concat([clusters_MCN, mapping_putative.reindex(clusters_MCN.index)], 1)\n",
    "\n",
    "lst = []\n",
    "for cl in sorted(clusters_MCN['cluster'].unique()):\n",
    "    temp = clusters_MCN[clusters_MCN['cluster'] == cl]['HMDB ID'].dropna().tolist()\n",
    "    for i in temp:\n",
    "        if i.count('HMDB') == 0:\n",
    "            pass\n",
    "        elif i.count('HMDB') > 1:\n",
    "            for j in i.replace(',','\\n').split('\\n'):\n",
    "                lst.append([j.replace('HMDB00','HMDB'), 'M-%d' % cl])\n",
    "        else:\n",
    "            lst.append([i.replace('\\n','').replace('HMDB00','HMDB'), 'M-%d' % cl])\n",
    "hmdb2cl = pd.DataFrame(lst, columns = ['index','cluster']).drop_duplicates()\n",
    "temp = clusters_GCN.reset_index()[['index', 'cluster']]\n",
    "temp['cluster'] = ['G-%d' % i for i in temp.cluster]\n",
    "\n",
    "cl_mapping = pd.concat([hmdb2cl, temp]).drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e848b8d0-1a33-4fd1-aa88-82ce102b270b",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_enrich = pd.Series()\n",
    "for cl in sorted(clusters_GCN['cluster'].unique()):\n",
    "    lst = clusters_GCN[clusters_GCN['cluster'] == cl].index.tolist()\n",
    "    enr = gp.enrichr(gene_list=lst,\n",
    "             gene_sets=GMT_genes,\n",
    "             background = tpm.shape[0],\n",
    "             outdir='test/enrichr_kegg2',\n",
    "             cutoff=0.5,\n",
    "             verbose=False)\n",
    "    temp = enr.res2d.sort_values('Term', ascending = True)#.iloc[0:5]\n",
    "    temp2 = temp[temp['P-value']<0.05][['Term','P-value']].set_index('Term')\n",
    "    temp2['gene'] = 'G-%d' % cl\n",
    "    gene_enrich = pd.concat([gene_enrich, temp2['gene']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "960bff32-617d-47ba-aa69-3a6fdf512225",
   "metadata": {},
   "outputs": [],
   "source": [
    "met_enrich = pd.Series()\n",
    "for cl in sorted(clusters_MCN['cluster'].unique()):\n",
    "    temp = clusters_MCN[clusters_MCN['cluster'] == cl]['HMDB ID'].dropna().tolist()\n",
    "    lst = []\n",
    "    for i in temp:\n",
    "        if i.count('HMDB') == 0:\n",
    "            pass\n",
    "        elif i.count('HMDB') > 1:\n",
    "            for j in i.replace(',','\\n').split('\\n'):\n",
    "                lst.append(j.replace('HMDB00','HMDB'))\n",
    "        else:\n",
    "            lst.append(i.replace('\\n','').replace('HMDB00','HMDB'))\n",
    "    y = []\n",
    "    for pw in list(GMT_mets.keys()):\n",
    "        hist_mets = set(lst).intersection(GMT_mets[pw])\n",
    "        hits = len(hist_mets)\n",
    "        num_ofpw = len(set(sum(list(GMT_mets.values()), [])))\n",
    "        num_ofmets_inpw = len(set(GMT_mets[pw]))\n",
    "        total_mets = len(set(lst))\n",
    "        pval = 1-hypergeom.cdf(hits,num_ofpw-num_ofmets_inpw,total_mets,num_ofmets_inpw)\n",
    "        y.append([pw, hits, num_ofmets_inpw, ';'.join(list(hist_mets)), pval])\n",
    "    y = pd.DataFrame(y, columns = ['Pathways', '# of Hits', 'Total Mets', 'Hits Metabolites', 'pval']).set_index('Pathways')\n",
    "    y = y[(y['# of Hits'] > 1) & (y['Total Mets'] > 1) & (y['pval'] < 0.05)]\n",
    "    y['met'] = 'M-%d' % cl\n",
    "    met_enrich = pd.concat([met_enrich, y['met']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a990f50f-9f22-4a5c-ba4a-5d4dba35bb7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "intersection = pd.DataFrame(gene_enrich).reset_index().merge(pd.DataFrame(met_enrich).reset_index(), on = 'index')\n",
    "## Use this file to generate chord diagram (Figure 6D)\n",
    "intersection[['0_y','0_x']].drop_duplicates().to_csv('Networks/ION/G2M.txt', sep = '\\t', index = False)\n",
    "\n",
    "## Combine the following files in Cytoscape to generate Figure 6E\n",
    "intersection[['index', '0_x']].to_csv('Networks/ION/G2PW.txt', sep = '\\t', index = False)\n",
    "intersection[['index', '0_y']].to_csv('Networks/ION/M2PW.txt', sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad2bd3fd-71c9-4d77-9a5a-20da6d6b2c70",
   "metadata": {},
   "outputs": [],
   "source": [
    "### R script used for figure 6D\n",
    "# library(circlize)\n",
    "\n",
    "# data = read.csv('Networks/ION/G2M.txt', sep = '\\t')\n",
    "# adjacencyData <- with(data, table(X0_y, X0_x))\n",
    "\n",
    "# grid.col = c(\n",
    "#   'M-0' = \"#17becf\", 'G-0' = \"#17becf\",\n",
    "#   'M-1' = \"#bcbd22\", 'G-1' = \"#bcbd22\",\n",
    "#   'M-2' = \"#7f7f7f\", 'G-2' = \"#7f7f7f\",\n",
    "#   'M-3' = \"#e377c2\", 'G-3' = \"#e377c2\",\n",
    "#   'G-4' = \"#8c564b\"\n",
    "# )\n",
    "\n",
    "# circos.par(start.degree = 90)\n",
    "# chordDiagram(adjacencyData, transparency = 0.25,\n",
    "#              annotationTrack = c(\"name\", \"grid\"),\n",
    "#              order = c('M-0', 'M-1', 'M-2', 'M-3',\n",
    "#                        'G-4', 'G-3', 'G-2', 'G-1', 'G-0'),\n",
    "#              grid.col = grid.col#, col = data$link\n",
    "# )\n",
    "# circos.clear()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f2edce5-2489-4d71-8551-15480123a073",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Figure 6B\n",
    "\n",
    "Use Supplementary Data 5\n",
    "\n",
    "OR\n",
    "\n",
    "1. Download the data from PRJNA358081, and PRJNA388978.\n",
    "2. Perform differential expression analysis using the code above\n",
    "\n",
    "Excluded outliers:\n",
    "* PRJNA358081: 'SRR5120909','SRR5120924','SRR5120930'\n",
    "* PRJNA388978: Use only the HC and IPFs\n",
    "\n",
    "To draw the Figure 6B, use the selected genes below (\"sel_genes\". Criteria: has to be significant in at least 2 timepoints and in all human IPF patients)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44de3d8b-0eb5-43b9-b1b7-31b26488f39a",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = pd.DataFrame()\n",
    "\n",
    "deseq = pd.ExcelFile('StatisticalAnalysis/Transcriptomics_TimeSeries/DifferentialExpression_ALL.xlsx')\n",
    "for i in deseq.sheet_names[0:]:\n",
    "    name = i.replace('WT_','').replace('Bleo_','Bleo-').replace('_',' vs ')\n",
    "    temp = deseq.parse(i, index_col = 0)\n",
    "    temp = temp[temp['P-ADJ'] < 0.05]['DIRECTION'].rename(name)\n",
    "    res = pd.concat([res, temp],1)\n",
    "res.index = res.index.str.upper()\n",
    "\n",
    "for i in supp5.sheet_names[0:3]:\n",
    "    name = i.replace('_',' vs ')\n",
    "    temp = supp5.parse(i, index_col = 0)\n",
    "    temp = temp[temp['P-ADJ'] < 0.05]['DIRECTION'].rename(name)\n",
    "    res = pd.concat([res, temp],1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f85c6a4-e6b9-41e5-be90-73fb7abc51a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_sel = res.reindex(res.iloc[0:, 0:-2].dropna(thresh = 2).index)\n",
    "sel_genes = res_sel[res_sel['GSE92592'].notna() & res_sel['GSE99621'].notna()]#.index"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd8a73cb-f929-4d45-9808-2b75f55f7d06",
   "metadata": {},
   "source": [
    "## Figure 6C-D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b92c70e7-56fa-498d-9422-c44561d09450",
   "metadata": {},
   "outputs": [],
   "source": [
    "clust = pd.read_csv('../Revision/Cluster_Human_new.txt', sep = '\\t', index_col = 0)['0'].replace(2,1)\n",
    "tpm = supp1.parse('TPM', index_col = 0)[metadata.index]\n",
    "\n",
    "KEGG = {i.split('\\t')[0]:[j for j in i.rstrip('\\n').split('\\t')[2:] if j != ''] for i in open('lib/KEGG_2019_Mouse.gmt', 'r').readlines()}\n",
    "exclude_KEGG = ['Hepatitis B', 'Human T-cell leukemia virus 1 infection', 'Regulation of actin cytoskeleton', 'Measles',\n",
    "           'MicroRNAs in cancer', 'Pathways in cancer', 'Influenza A', 'Oocyte meiosis', 'Progesterone-mediated oocyte maturation',\n",
    "           'Ribosome biogenesis in eukaryotes', 'Amoebiasis', 'Small cell lung cancer', 'Epstein-Barr virus infection', 'Human papillomavirus infection',\n",
    "           'Pertussis', 'Staphylococcus aureus infection', 'Proteoglycans in cancer', 'Leishmaniasis', 'Legionellosis', 'Viral carcinogenesis', \n",
    "           'Chagas disease (American trypanosomiasis)', 'Toxoplasmosis', 'Malaria', 'Metabolism of xenobiotics by cytochrome P450', 'Chronic myeloid leukemia',\n",
    "           'Prion diseases', 'Inflammatory bowel disease (IBD)', 'Kaposi sarcoma-associated herpesvirus infection', 'Parkinson disease', 'Hepatitis C', 'Systemic lupus erythematosus',\n",
    "           'Tuberculosis', 'Hepatitis B', 'Measles', 'Dilated cardiomyopathy (DCM)', 'Hematopoietic cell lineage', 'Adrenergic signaling in cardiomyocytes',\n",
    "           'Epstein-Barr virus infection', 'Drug metabolism', 'Salivary secretion', 'Thermogenesis', 'Glioma', 'Human cytomegalovirus infection', 'Melanoma', 'Chemical carcinogenesis',\n",
    "            'African trypanosomiasis', 'Collecting duct acid secretion', 'African trypanosomiasis', 'Vascular smooth muscle contraction', 'Fluid shear stress and atherosclerosis',\n",
    "            'Rheumatoid arthritis', 'Allograft rejection', 'Long term depression'\n",
    "          ]\n",
    "\n",
    "exclude_GO = ['',\n",
    "             \n",
    "             ]\n",
    "\n",
    "color_mapping = dict(zip(sorted(clust.unique()),sns.color_palette(\"tab10\").as_hex()))\n",
    "\n",
    "\n",
    "fig, axs = plt.subplots(figsize = (2, 3),nrows = len(clust.unique().tolist()), ncols = 1, sharex=False\n",
    "                        , gridspec_kw={'height_ratios':[5, 5]})\n",
    "\n",
    "for j, sel_ax in enumerate(axs):\n",
    "    print()\n",
    "    gl = [ha.upper() for ha in clust[clust == j].index.tolist()]\n",
    "    enr = gp.enrichr(gene_list=gl,\n",
    "             gene_sets=KEGG,\n",
    "             background = tpm.shape[0],\n",
    "             outdir='test/enrichr_kegg2',\n",
    "             cutoff=1,\n",
    "             verbose=False)\n",
    "    temp = enr.res2d[~(\n",
    "        enr.res2d['Term'].isin(exclude_KEGG) |\n",
    "        (enr.res2d['Term'].str.contains('isease')) |\n",
    "         (enr.res2d['Term'].str.contains('arcinoma')) |\n",
    "          (enr.res2d['Term'].str.contains('ancer')) |\n",
    "        (enr.res2d['Term'].str.contains('thyroid')) |\n",
    "        (enr.res2d['Term'].str.contains('yndrome')) |\n",
    "        (enr.res2d['Term'].str.contains('strogen')) |\n",
    "        (enr.res2d['Term'].str.contains('elanoge')) |\n",
    "        (enr.res2d['Term'].str.contains('iabet')) |\n",
    "        (enr.res2d['Term'].str.contains('ysosome')) |\n",
    "        (enr.res2d['Term'].str.contains('nfection')) |\n",
    "        (enr.res2d['Term'].str.contains('ntestin')) |\n",
    "        (enr.res2d['Term'].str.contains('cardi')) |\n",
    "        (enr.res2d['Term'].str.contains('mmunodef'))\n",
    "    )].sort_values('Adjusted P-value', ascending = True)#.iloc[0:5]#.iloc[[2,3,4]]['Term'].tolist()\n",
    "    temp2 = temp[['Term','P-value','Adjusted P-value']]\n",
    "    temp2 = temp2[temp2['Adjusted P-value'] < 0.05]\n",
    "    print(temp2.shape)\n",
    "    temp2['Term'] = temp2['Term'].str.replace('endoplasmic reticulum', 'ER')\n",
    "    temp2['-Log10(P-Adj)'] = -np.log10(temp2['Adjusted P-value'])\n",
    "    sns.barplot(data = temp2, x = '-Log10(P-Adj)', y = 'Term', ax = sel_ax, color = color_mapping[j])\n",
    "    sel_ax.yaxis.tick_right()\n",
    "    sel_ax.set_yticklabels([i for i in temp2['Term']], fontsize = 15)\n",
    "#     sel_ax.set_xticks([])\n",
    "    sel_ax.set_ylabel('DEG-%d (%d Genes)' % (j, len(gl)), fontsize = 15, fontweight = 'bold')\n",
    "    sel_ax.tick_params(right = False)\n",
    "    if j == clust.max():\n",
    "        sns.despine(ax = sel_ax)\n",
    "        sel_ax.set_xlabel('-Log10(P-Adj)', fontsize = 15, fontweight = 'bold')\n",
    "    else:\n",
    "        sns.despine(ax = sel_ax, bottom = False)\n",
    "        sel_ax.set_xlabel('')\n",
    "plt.savefig('Figures/Figure6C-D_funct.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85edc8af-1650-49b0-b828-029ec9d45b90",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Figure 7A, C, D\n",
    "\n",
    "This was done via iDREM. Please refer to iDREM tutorial and Supplementary Table 6\n",
    "\n",
    "Figure 7C (and Supplementary Figure 5) was regenerated using Pysankey package, but can be generated directly using iDREM.\n",
    "\n",
    "Final iDREM model and visualization can be found in the github repository"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "033643bd-55c6-4660-8d3e-70e88a1ed0be",
   "metadata": {},
   "source": [
    "## Figure 7B\n",
    "\n",
    "Use \"iDREM_Results/GCN_iDREM_edges.txt\" generated below as the input to cytoscape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f3449f5-a950-4990-848f-daf3b29271ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "k = pickle.load(open('Networks/ReadyObjects/GCN.pkl', 'rb'))\n",
    "clust = k.clustering\n",
    "clust.index = clust.index.str.upper()\n",
    "iDREM_mapping = supp6.parse('Gene Tracks', index_col = 0)['Tracks']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81df5da3-c1d1-4359-91e5-5d36db1ef8f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "num_ofpw = clust.shape[0]\n",
    "y = []\n",
    "for cl in sorted(clust.unique()):\n",
    "    temp = clust[clust == cl]\n",
    "    for node in iDREM_mapping.unique():\n",
    "        hits_an = set(temp.index.str.upper()).intersection(iDREM_mapping[iDREM_mapping == node].index)\n",
    "        hits = len(hits_an)\n",
    "        num_member_inpw = iDREM_mapping[iDREM_mapping == node].shape[0]\n",
    "        pval = 1-hypergeom.cdf(hits,num_ofpw-num_member_inpw, temp.shape[0],num_member_inpw)\n",
    "        y.append([node, 'G-%s' % cl, pval])\n",
    "results_cl = pd.DataFrame(y, columns = ['node', 'cl', 'pval'])\n",
    "results_cl['padj']=multipletests(results_cl['pval'],method='fdr_bh')[1]\n",
    "results_cl = results_cl[results_cl['padj'] < 0.05].sort_values(['node'], ascending = False)\n",
    "results_cl['-Log10(P-Adj)'] = -np.log1p(results_cl['padj'])\n",
    "results_cl.to_csv('iDREM_Results/GCN_iDREM_edges.txt', sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21ffcb60-8c31-4f73-afd3-90442e749d69",
   "metadata": {},
   "source": [
    "## Differential Expression & Functional Analysis of CB1R KO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26cdfa7b-a204-4f7f-b4bc-305d1aec9686",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = supp5.parse('Metadata and PFTs', index_col = 1)\n",
    "metadata = temp.iloc[0:, 0:2]\n",
    "count = supp5.parse('Counts', index_col = 0)[metadata.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3d69b44-eff6-4fa1-b2c9-390fdedd9544",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_path='StatisticalAnalysis/Transcriptomics_CB1RKO/'\n",
    "k=DESeq_Piano(res_path)\n",
    "k.DEseq(count,metadata['Study group'].str.replace('+', '_').str.replace(' ', '').str.replace('-', ''))\n",
    "k.loadGSC(KEGG='lib/KEGG_2019_Mouse.gmt')\n",
    "\n",
    "cond2='Control'\n",
    "cond1='wildtype_Bleomycin'\n",
    "k.DEseq_Compare(cond1,cond2,save=True)\n",
    "k.PIANO(cond1,cond2,save=True)\n",
    "\n",
    "cond2='wildtype_Bleomycin'\n",
    "cond1='GlobalCB1RKO_Bleomycin'\n",
    "k.DEseq_Compare(cond1,cond2,save=True)\n",
    "k.PIANO(cond1,cond2,save=True)\n",
    "\n",
    "k.summarizeDEseq()\n",
    "k.summarizePiano()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c77223e1-1d8c-458d-add5-11ff9a6adcaa",
   "metadata": {},
   "source": [
    "## Figure 8A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0671d63f-55ca-40c2-bbaf-ad31036c3b10",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = supp7.parse('Metadata and PFTs', index_col = 1)\n",
    "metadata = temp.iloc[0:, 0:1]\n",
    "PFT = temp.iloc[0:, 1:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5db7cba-876e-4539-b3c3-33adccd7cfb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(figsize = (13, 2.5),nrows = 1, ncols = 5, sharex=True)\n",
    "# axs = sum([list(i) for i in axs], [])\n",
    "\n",
    "for num, pft in enumerate(PFT.columns[1:][::-1]):\n",
    "    temp = pd.concat([PFT[pft], metadata],1).reset_index()\n",
    "    order = ['Control','wild-type+ Bleomycin', ' Global CB1R KO + Bleomycin']\n",
    "    box_pairs = [\n",
    "        ('Control', 'wild-type+ Bleomycin'),\n",
    "        ('Control', ' Global CB1R KO + Bleomycin'),\n",
    "        (' Global CB1R KO + Bleomycin', 'wild-type+ Bleomycin')\n",
    "    ]\n",
    "    \n",
    "    pval_sel = [\n",
    "        f_oneway(temp[temp['Study group'] == 'wild-type+ Bleomycin'][pft], \n",
    "             temp[temp['Study group'] == 'Control'][pft])[1],\n",
    "        f_oneway(temp[temp['Study group'] == 'Control'][pft], \n",
    "             temp[temp['Study group'] == ' Global CB1R KO + Bleomycin'][pft])[1],\n",
    "        f_oneway(temp[temp['Study group'] == 'wild-type+ Bleomycin'][pft],\n",
    "             temp[temp['Study group'] == ' Global CB1R KO + Bleomycin'][pft])[1]\n",
    "    ]\n",
    "\n",
    "    sns.boxplot(data = temp, x = 'Study group', y = pft, order=order, ax = axs[num])\n",
    "    add_stat_annotation(data = temp, x = 'Study group', y = pft, order=order, ax = axs[num],\n",
    "                        box_pairs=box_pairs,\n",
    "                        perform_stat_test=False, pvalues=pval_sel,\n",
    "                        test=None, text_format='star', loc='inside', verbose=0, fontsize = 15)\n",
    "    axs[num].set_xticklabels(axs[num].get_xticklabels(), fontsize = 15, fontweight = 'bold', rotation = 30, ha = 'right')\n",
    "    axs[num].set_xlabel('')\n",
    "    axs[num].set_ylabel('')\n",
    "    if num < 3:\n",
    "        axs[num].set_yticks([0.5, 1, 1.5])\n",
    "    axs[num].set_yticklabels(['%d' % j if int(j) == j else '%.1f' % j for j in axs[num].get_yticks()], fontsize = 15)\n",
    "    axs[num].set_title('%s' % (pft.replace(' (PFT)', '')), fontsize = 15, fontweight = 'bold')\n",
    "    plt.subplots_adjust(wspace = 0.3)\n",
    "plt.savefig('Figures/Figure8A.pdf', transparent = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b107e4cf-b672-49bb-8ff4-36ca16a91112",
   "metadata": {},
   "source": [
    "## Figure 8B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "512799a2-2f49-43ee-a984-5404a039ebc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "hyp = supp7.parse('Hydroxyproline')\n",
    "qpcr = supp7.parse('qPCR', index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "932816db-01a6-4095-8f3e-30354ed0f76d",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(figsize = (13, 2.5),nrows = 1, ncols = 5, sharex=True)\n",
    "\n",
    "order = ['Control','wild-type+ Bleomycin', ' Global CB1R KO + Bleomycin']\n",
    "box_pairs = [\n",
    "    ('Control', 'wild-type+ Bleomycin'),\n",
    "    ('Control', ' Global CB1R KO + Bleomycin'),\n",
    "    (' Global CB1R KO + Bleomycin', 'wild-type+ Bleomycin')\n",
    "]\n",
    "\n",
    "var = 'Hydroxyproline'\n",
    "num = 0\n",
    "\n",
    "pval_sel = [\n",
    "    f_oneway(hyp[hyp['Unnamed: 0'] == 'wild-type+ Bleomycin']['Hydroxyproline'], \n",
    "         hyp[hyp['Unnamed: 0'] == 'Control']['Hydroxyproline'])[1],\n",
    "    f_oneway(hyp[hyp['Unnamed: 0'] == 'Control']['Hydroxyproline'], \n",
    "         hyp[hyp['Unnamed: 0'] == ' Global CB1R KO + Bleomycin']['Hydroxyproline'])[1],\n",
    "    f_oneway(hyp[hyp['Unnamed: 0'] == 'wild-type+ Bleomycin']['Hydroxyproline'],\n",
    "         hyp[hyp['Unnamed: 0'] == ' Global CB1R KO + Bleomycin']['Hydroxyproline'])[1]\n",
    "]\n",
    "sns.boxplot(data = hyp, x = 'Unnamed: 0', y = var, order=order, ax = axs[num])\n",
    "add_stat_annotation(data = hyp, x = 'Unnamed: 0', y = var, order=order, ax = axs[num],\n",
    "                    box_pairs=box_pairs,\n",
    "                    perform_stat_test=False, pvalues=pval_sel,\n",
    "                    test=None, text_format='star', loc='inside', verbose=0, fontsize = 15)\n",
    "axs[num].set_xticklabels(axs[num].get_xticklabels(), fontsize = 15, fontweight = 'bold', rotation = 30, ha = 'right')\n",
    "axs[num].set_xlabel('')\n",
    "axs[num].set_ylabel('')\n",
    "axs[num].set_yticklabels(['%d' % j if int(j) == j else '%.1f' % j for j in axs[num].get_yticks()], fontsize = 15)\n",
    "axs[num].set_title('%s' % var, fontsize = 15, fontweight = 'bold')\n",
    "plt.subplots_adjust(wspace = 0.3)\n",
    "\n",
    "\n",
    "for num, var in enumerate(qpcr.columns[1:-1]):\n",
    "    num = num + 1\n",
    "    fc = pd.concat([qpcr[['conds']+[var]]],1).dropna()\n",
    "    order = ['Control','WT-Bleo', 'CB1R KO-Bleo']\n",
    "    box_pairs = [\n",
    "        ('Control', 'WT-Bleo'),\n",
    "        ('CB1R KO-Bleo', 'Control'),\n",
    "        ('CB1R KO-Bleo', 'WT-Bleo')\n",
    "                ]\n",
    "    \n",
    "    pval_sel = [\n",
    "        f_oneway(fc[fc['conds'] == 'WT-Bleo'][var], \n",
    "             fc[fc['conds'] == 'Control'][var])[1],\n",
    "        f_oneway(fc[fc['conds'] == 'Control'][var], \n",
    "             fc[fc['conds'] == 'CB1R KO-Bleo'][var])[1],\n",
    "        f_oneway(fc[fc['conds'] == 'WT-Bleo'][var], \n",
    "             fc[fc['conds'] == 'CB1R KO-Bleo'][var])[1],\n",
    "    ]\n",
    "\n",
    "    sns.boxplot(data = fc, x = 'conds', y = var, order=order, ax = axs[num])\n",
    "    add_stat_annotation(data = fc, x = 'conds', y = var, order=order, ax = axs[num],\n",
    "                        box_pairs=box_pairs,\n",
    "                        perform_stat_test=False, pvalues=pval_sel,\n",
    "                        test=None, text_format='star', loc='inside', verbose=0, fontsize = 15)\n",
    "    axs[num].set_xticklabels(axs[num].get_xticklabels(), fontsize = 15, fontweight = 'bold', rotation = 30, ha = 'right')\n",
    "    axs[num].set_xlabel('')\n",
    "    axs[num].set_ylabel('')\n",
    "    if num == 3:\n",
    "        axs[num].set_yticks([1, 2])\n",
    "        axs[num].set_ylim([0.75, 2.5])\n",
    "    axs[num].set_yticklabels(['%d' % j if int(j) == j else '%.1f' % j for j in axs[num].get_yticks()], fontsize = 15)\n",
    "    axs[num].set_title('%s' % var, fontsize = 15, fontweight = 'bold')\n",
    "    plt.subplots_adjust(wspace = 0.3)\n",
    "plt.savefig('Figures/Figure8B-C.pdf', transparent = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9017a2a-719c-4db3-9b55-a56657553a53",
   "metadata": {},
   "source": [
    "## Figure 8D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07c1b62d-5b5e-4c0a-b5a5-4125d07b92e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "piano = pd.ExcelFile('StatisticalAnalysis/Transcriptomics_TimeSeries/KEGG_ALL.xlsx')\n",
    "res = pd.DataFrame()\n",
    "for i in ['Bleomycin_14Days_Control']:\n",
    "    temp = piano.parse(i, index_col = 0)\n",
    "    temp = temp['Stats'].rename(i)\n",
    "    res = pd.concat([res, temp],1)\n",
    "\n",
    "piano = pd.ExcelFile('StatisticalAnalysis/Transcriptomics_CB1RKO/KEGG_ALL.xlsx')\n",
    "for i in piano.sheet_names[0:]:\n",
    "    temp = piano.parse(i, index_col = 0)\n",
    "    temp = temp['Stats'].rename(i)\n",
    "    res = pd.concat([res, temp],1)\n",
    "    \n",
    "res = res[[\n",
    "    'Bleomycin_14Days_Control', 'wildtype_Bleomycin_Control', 'GlobalCB1RKO_Bleomycin_wildtype'\n",
    "]]\n",
    "    \n",
    "res = res[[\n",
    "    'Bleomycin_14Days_Control', 'wildtype_Bleomycin_Control', 'GlobalCB1RKO_Bleomycin_wildtype'\n",
    "]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "622cae26-edc0-4a12-acad-579ccc3f8579",
   "metadata": {},
   "outputs": [],
   "source": [
    "exclude_KEGG = ['Hepatitis B', 'Human T-cell leukemia virus 1 infection', 'Regulation of actin cytoskeleton', 'Measles',\n",
    "           'MicroRNAs in cancer', 'Pathways in cancer', 'Influenza A', 'Oocyte meiosis', 'Progesterone-mediated oocyte maturation',\n",
    "           'Ribosome biogenesis in eukaryotes', 'Amoebiasis', 'Small cell lung cancer', 'Epstein-Barr virus infection', 'Human papillomavirus infection',\n",
    "           'Pertussis', 'Staphylococcus aureus infection', 'Proteoglycans in cancer', 'Leishmaniasis', 'Legionellosis', 'Viral carcinogenesis', \n",
    "           'Chagas disease (American trypanosomiasis)', 'Toxoplasmosis', 'Malaria', 'Metabolism of xenobiotics by cytochrome P450', 'Chronic myeloid leukemia',\n",
    "           'Prion diseases', 'Inflammatory bowel disease (IBD)', 'Kaposi sarcoma-associated herpesvirus infection', 'Parkinson disease', 'Hepatitis C', 'Systemic lupus erythematosus',\n",
    "           'Tuberculosis', 'Hepatitis B', 'Measles', 'Dilated cardiomyopathy (DCM)', 'Hematopoietic cell lineage', 'Adrenergic signaling in cardiomyocytes',\n",
    "           'Epstein-Barr virus infection', 'Drug metabolism', 'Salivary secretion', 'Thermogenesis', 'Glioma', 'Human cytomegalovirus infection', 'Melanoma', 'Chemical carcinogenesis',\n",
    "            'African trypanosomiasis', 'Collecting duct acid secretion', 'African trypanosomiasis', 'Vascular smooth muscle contraction', 'Fluid shear stress and atherosclerosis',\n",
    "            'Rheumatoid arthritis', 'Allograft rejection', 'Long term depression'\n",
    "          ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "535bc37b-a6d3-4448-989a-d88d805c8fec",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_sel = res[~(\n",
    "    res.index.isin(exclude_KEGG) |\n",
    "    (res.index.str.contains('isease')) |\n",
    "     (res.index.str.contains('arcinoma')) |\n",
    "      (res.index.str.contains('ancer')) |\n",
    "    (res.index.str.contains('thyroid')) |\n",
    "    (res.index.str.contains('yndrome')) |\n",
    "    (res.index.str.contains('strogen')) |\n",
    "    (res.index.str.contains('elanoge')) |\n",
    "    (res.index.str.contains('iabet')) |\n",
    "    (res.index.str.contains('ysosome')) |\n",
    "    (res.index.str.contains('nfection')) |\n",
    "    (res.index.str.contains('ntestin')) |\n",
    "    (res.index.str.contains('cardi')) |\n",
    "    (res.index.str.contains('Cardi')) |\n",
    "    (res.index.str.contains('nemia')) |\n",
    "    (res.index.str.contains('mmunodef'))\n",
    ")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5a0de41-e5f5-4c1b-a437-334b6b67772b",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = res_sel.dropna()\n",
    "cmap=matplotlib.colors.LinearSegmentedColormap.from_list(\"\", [\"#0000a5\",'#0000d8',\"#FFFAF0\",'#d80000',\"#a50000\"])\n",
    "g = sns.clustermap(temp.fillna(0), yticklabels = 1, \n",
    "               center = 0, cmap = cmap, \n",
    "               col_cluster = False, row_cluster = True,\n",
    "               figsize = (7,10),\n",
    "               cbar_kws={\"orientation\": \"vertical\",\"shrink\": 1, 'label': 'Gene-Set Statistic'}\n",
    "              )\n",
    "g.ax_heatmap.set_xticklabels(g.ax_heatmap.get_xticklabels(),rotation=30, ha = 'right', fontsize = 15, fontweight = 'bold')\n",
    "g.ax_heatmap.set_yticklabels(g.ax_heatmap.get_yticklabels(), fontsize = 15, )\n",
    "g.ax_heatmap.tick_params(right=False, bottom=False)\n",
    "g.savefig('Figures/Figure8C.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42ef26f5-853c-4392-8fcf-c2a5fed508d3",
   "metadata": {},
   "source": [
    "## Figure 8E\n",
    "\n",
    "Direction for Bleo+CB1R-KO was based on the hypergeometric test between the clusters and differential expression analysis of CB1R KO vs WildType Bleomycin, as calculated below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f898463a-cef0-4968-96b6-e07cd77f6c17",
   "metadata": {},
   "outputs": [],
   "source": [
    "k = pickle.load(open('Networks/ReadyObjects/GCN.pkl', 'rb'))\n",
    "clust = k.clustering\n",
    "tpm = supp7.parse('TPM', index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d801f798-8359-429b-ad24-ebf6a5a18372",
   "metadata": {},
   "outputs": [],
   "source": [
    "deseq = pd.ExcelFile('StatisticalAnalysis/Transcriptomics_CB1RKO/DifferentialExpression_ALL.xlsx')\n",
    "\n",
    "res = pd.DataFrame()\n",
    "for i in deseq.sheet_names:\n",
    "    temp = deseq.parse(i, index_col = 0)\n",
    "    temp = temp[temp['P-ADJ'] < 0.05]['DIRECTION'].rename(i)\n",
    "    res = pd.concat([res, temp],1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f6afe25-9061-4ce2-8451-c301ff2856a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "num_ofpw = tpm.shape[0]\n",
    "y = []\n",
    "for cl in sorted(clust.unique()):\n",
    "    temp = clust[clust == cl]\n",
    "    for direction in ['UP','DOWN']:\n",
    "        hits_an = set(temp.index).intersection(res[res['GlobalCB1RKO_Bleomycin_wildtype'] == direction].index)\n",
    "        hits = len(hits_an)\n",
    "        num_member_inpw = res[res['GlobalCB1RKO_Bleomycin_wildtype'] == direction].shape[0]\n",
    "        pval = 1-hypergeom.cdf(hits,num_ofpw-num_member_inpw, temp.shape[0],num_member_inpw)\n",
    "        y.append([direction, 'G-%d' % cl, pval])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "109790b8-a866-4302-ad1b-5f58219312a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_cl = pd.DataFrame(y, columns = ['node', 'cl', 'pval'])\n",
    "results_cl['padj']=multipletests(results_cl['pval'],method='fdr_bh')[1]\n",
    "results_cl = results_cl[results_cl['padj'] < 0.05].sort_values(['node'], ascending = False)\n",
    "results_cl"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14da2532-862a-48de-8682-f9d374ed9166",
   "metadata": {},
   "source": [
    "## Figure 7D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f57faca-220c-4ba1-acec-794aa8a8d1ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = pd.DataFrame()\n",
    "\n",
    "for i in ['DEGs 14 Days']:\n",
    "    name = i.replace('WT_','').replace('Bleo_','Bleo-').replace('_',' vs ')\n",
    "    temp = supp1.parse(i, index_col = 0)\n",
    "    temp = temp[temp['P-ADJ'] < 0.05]['L2FC'].rename(name)\n",
    "    res = pd.concat([res, temp],1)\n",
    "res.index = res.index.str.upper()\n",
    "\n",
    "for i in supp5.sheet_names:\n",
    "        name = i.replace('_',' vs ')\n",
    "        temp = supp5.parse(i, index_col = 0)\n",
    "        temp = temp[temp['P-ADJ'] < 0.05]['L2FC'].rename(name)\n",
    "        res = pd.concat([res, temp],1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd510564-8d02-4589-a6da-1ab6c03c4e10",
   "metadata": {},
   "outputs": [],
   "source": [
    "iDREM_mapping = supp6.parse('Gene Tracks', index_col = 0)['Tracks']\n",
    "gn = res[np.sign(res).sum(1).abs() == 3].dropna(how = 'all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0ed7ff2-5a8d-4ea1-a629-dab448e448f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "lut = dict(zip(iDREM_mapping.unique(), plt.cm.tab20.colors))\n",
    "row_colors = iDREM_mapping.map(lut)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a87519b-7282-425b-a25c-176420f79e8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "cmap=matplotlib.colors.LinearSegmentedColormap.from_list(\"\", [\"#0000a5\",'#0000d8',\"#FFFAF0\",'#d80000',\"#a50000\"])\n",
    "temp = gn[gn.index.isin(iDREM_mapping[iDREM_mapping == 'A'].index)]\n",
    "g = sns.clustermap(temp, cmap = cmap, center = 0, figsize = (3, 5), yticklabels = 1, col_cluster = False, vmin = 0, vmax = 8.2, row_colors = row_colors)\n",
    "g.savefig('Figures/7E_subsetA.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "576dbd07-5718-4fa8-b957-9c6b01fb1c25",
   "metadata": {},
   "outputs": [],
   "source": [
    "cmap=matplotlib.colors.LinearSegmentedColormap.from_list(\"\", [\"#0000a5\",'#0000d8',\"#FFFAF0\",'#d80000',\"#a50000\"])\n",
    "temp = gn[gn.index.isin(iDREM_mapping[iDREM_mapping == 'B'].index)]\n",
    "g = sns.clustermap(temp, cmap = cmap, center = 0, figsize = (3, 12), yticklabels = 1, col_cluster = False, vmin = 0, vmax = 8.2, row_colors = row_colors)\n",
    "g.ax_heatmap.set_xticks([])\n",
    "g.savefig('Figures/7E_subsetB.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25100ab2-8edc-4865-9c54-70f0a530fb98",
   "metadata": {},
   "source": [
    "## Figure 8H\n",
    "(Execute Figure 7D first, as this is subsetting the genes there)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "877e4733-816c-44f9-9207-ff29db302e2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = pd.DataFrame()\n",
    "\n",
    "for i in ['DEGs Bleo WT vs Control', 'DEGs Bleo KO vs Bleo WT']:\n",
    "    name = i.replace('WT_','').replace('Bleo_','Bleo-').replace('_',' vs ')\n",
    "    temp = supp7.parse(i, index_col = 0)\n",
    "    temp = temp[temp['P-VALUE'] < 0.05]['L2FC'].rename(name)\n",
    "    res = pd.concat([res, temp],1)\n",
    "res.index = res.index.str.upper()\n",
    "\n",
    "for i in supp5.sheet_names:\n",
    "        name = i.replace('_',' vs ')\n",
    "        temp = supp5.parse(i, index_col = 0)\n",
    "        temp = temp[temp['P-ADJ'] < 0.05]['L2FC'].rename(name)\n",
    "        res = pd.concat([res, temp],1)\n",
    "\n",
    "res = res.reindex(gn.index).dropna()\n",
    "\n",
    "res = res[['GSE92592', 'GSE99621', 'DEGs Bleo WT vs Control', 'DEGs Bleo KO vs Bleo WT']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4048600-73bb-4ef3-b010-463c6c30fd01",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = res[res.index.isin(iDREM_mapping.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bec36e70-2fce-4bda-97bd-498f6657c200",
   "metadata": {},
   "outputs": [],
   "source": [
    "lut = dict(zip(iDREM_mapping.unique(), plt.cm.tab20.colors))\n",
    "row_colors = iDREM_mapping.map(lut)\n",
    "\n",
    "cmap=matplotlib.colors.LinearSegmentedColormap.from_list(\"\", [\"#0000a5\",'#0000d8',\"#FFFAF0\",'#d80000',\"#a50000\"])\n",
    "temp = res[res.index.isin(iDREM_mapping[iDREM_mapping == 'A'].index)]\n",
    "\n",
    "g = sns.clustermap(temp, cmap = cmap, center = 0, figsize = (4, 4), yticklabels = 1, col_cluster = False, vmin = -1.7, vmax = 6.5, row_colors = row_colors)\n",
    "g.ax_heatmap.set_xticks([])\n",
    "g.cax.set_visible(False)\n",
    "g.savefig('Figures/8H_subsetA.pdf')\n",
    "\n",
    "cmap=matplotlib.colors.LinearSegmentedColormap.from_list(\"\", [\"#0000a5\",'#0000d8',\"#FFFAF0\",'#d80000',\"#a50000\"])\n",
    "temp = res[res.index.isin(iDREM_mapping[iDREM_mapping == 'B'].index)]\n",
    "\n",
    "g = sns.clustermap(temp, cmap = cmap, center = 0, figsize = (4, 10), yticklabels = 1, col_cluster = False, vmin = -1.7, vmax = 6.5, row_colors = row_colors)\n",
    "g.ax_heatmap.set_xticks([])\n",
    "g.cax.set_visible(False)\n",
    "g.savefig('Figures/8H_subsetB.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd312570-8b10-46a7-aa6f-b5f2af0331b2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "PFProgression_Analysis",
   "language": "python",
   "name": "pfprogression_analysis"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  },
  "toc-autonumbering": false,
  "toc-showcode": false,
  "toc-showmarkdowntxt": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
